{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### General intro"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Thanks to the rapid advances in sequencing technologies generating genome-wide sequence datasets for any species has become routine and there is great interest in learning about the demographic and selective past of populations from sequence variation. The coalescent gives an elegant mathematical description of the ancestry of a sample of sequences from a more or less idealized population and, given its focus on samples, has become the bread and butter of modern population genetics. \n",
    "\n",
    "However, despite the flood of sequence data and the plethora of coalescent-based inference tools now available, many analyses of genome wide variation remain superficial or entirely descriptive. Developing efficient inference methods has been difficult for two reasons. First, analytic results for models of population structure and/or history are often restricted to average coalescence times and small (often pairwise) samples. Even when it is possible to derive the full distribution of genealogies for realistic samples sizes and models, the results are cumbersome and generally rely on automation (Lohse 2016). Second, and perhaps more fundamentally, dealing with recombination has proven extremely challenging and we still lack a full analytic description of genetic drift and recombination even under the simplest null models (in the absence of demography and selection). Thus, inference methods that incorporate linkage information generally rely on drastic simplifying assumptions about recombination (cite SMC papers and Song).\n",
    "\n",
    "Given the severe limitations of relating sequence variation to mechanistic models analytically, coalescent simulations have become an integral part of inference in a number of ways: First, comparisons between analytic results against simulations serve as an important sanity check for both. Second, while it is often possible to use analytic approaches to obtain unbiased point estimates of demographic parameters by ignoring linkage (cite dadi and ), quantifying the uncertainty and potential biases in such estimates requires simulating data with linkage to conduct parametric bootstrapping. Finally, a range of inference approaches that rely on coalescent simulations directly to approximate the likelihood or - in a Bayesian setting - the posterior, of parameters under arbitrarily complex models of demography: Approximate Bayesian computation (ABC) (cite DIYABC, msBayes) has evolved ii) Approximate likelihood methods either based on SNPs (cite Excoffier) or blockwise data (cite ABLE). \n",
    "\n",
    "From the perspective of the user, inference tools are generally implemented with a command line or graphical user interface and are designed for a more or less narrow set of inference problems. Moreover, curent inference methods are based on a variety of coalescent simulators which are often. The aim of this chapter is to illustrate that how msprime (and its API) provides a flexible way to run and use coalescent simulations under under any model. We first explain the basic datastruture and methods msprimes for coalescent simulations and then show by the way of a set of simple practical examples how msprime can be used to obtain predictions for commonly used data summaries and diffferent models of demography. Finally, we show how msprime can be used to set up a ABC inference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format = 'svg'\n",
    "import msprime\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import multiprocessing\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from IPython.display import SVG\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running simulations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At the simplest level, coalescent simulation is about generating trees. These trees represent the simulated history of a sample of individuals drawn from an idealised population (in later sections we show how to vary the properites of this idealised population). The function ``msprime.simulate`` runs these simulations, and the parameters that we provide to this function determine the exact simulation that we run. It returns a ``TreeSequence`` object, which represents the full simulated history of the sample. In later sections we discuss the effects of recombination, when this ``TreeSequence`` contains a list of correlated trees. For now, we assume that there is no recombinations, and we use the method ``first()`` to obtain the tree object from this tree sequence. For example, here we simulate a history for a sample of three chromosomes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs/><g id=\"lines\" stroke=\"black\"><line x1=\"40.0\" x2=\"40.0\" y1=\"180.0\" y2=\"20.0\"/><line x1=\"40.0\" x2=\"70.0\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"100.0\" x2=\"100.0\" y1=\"103.8333704157443\" y2=\"20.0\"/><line x1=\"100.0\" x2=\"70.0\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"80.0\" x2=\"80.0\" y1=\"180.0\" y2=\"103.8333704157443\"/><line x1=\"80.0\" x2=\"100.0\" y1=\"103.8333704157443\" y2=\"103.8333704157443\"/><line x1=\"120.0\" x2=\"120.0\" y1=\"180.0\" y2=\"103.8333704157443\"/><line x1=\"120.0\" x2=\"100.0\" y1=\"103.8333704157443\" y2=\"103.8333704157443\"/></g><g font-size=\"14\" text-anchor=\"start\"><text x=\"105.0\" y=\"98.8333704157443\">3</text></g><g font-size=\"14\" text-anchor=\"end\"/><g font-size=\"14\" text-anchor=\"middle\"><text x=\"70.0\" y=\"15.0\">4</text><text x=\"40.0\" y=\"200.0\">0</text><text x=\"80.0\" y=\"200.0\">1</text><text x=\"120.0\" y=\"200.0\">2</text></g><circle cx=\"70.0\" cy=\"20.0\" fill=\"black\" r=\"3\"/><circle cx=\"40.0\" cy=\"180.0\" fill=\"black\" r=\"3\"/><circle cx=\"100.0\" cy=\"103.8333704157443\" fill=\"black\" r=\"3\"/><circle cx=\"80.0\" cy=\"180.0\" fill=\"black\" r=\"3\"/><circle cx=\"120.0\" cy=\"180.0\" fill=\"black\" r=\"3\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"start\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"end\"/></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.simulate(3)\n",
    "tree = ts.first()\n",
    "SVG(tree.draw())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this tree we have 5 nodes. Nodes 0, 1 and 2 are *leaves*, and represent our samples. Node 3 is an *internal* node, and is the parent of 0 and 1. Node 4 is also an internal node, and is the root of the tree. In mpsrime, we always refer to nodes by their integer IDs and obtain information about these nodes by calling methods on the tree object. For example the code ``tree.children(4)`` will return the tuple ``(2, 3)`` here, as these are the node IDs of the children of the root node. Similarly, ``tree.parent(0)`` will return ``3``.\n",
    "\n",
    "The height of nodes in a tree are determined by their *time*; this is how long ago the ancestor that corresponds the node was born. So, contemporary samples always have a node time of zero, and time values increase as we go upwards in the tree (and further back in time). Times in msprime are always measured in *generations*. \n",
    "\n",
    "When we run a single simulation, the resulting tree is a single random sample from the probability distribution of coalescent trees. Since a single random draw from any distribution is not usually informative, we nearly always need to run many different *replicate* simulations to obtain useful information. This simplest way to do this in msprime is to use the ``num_replicates`` argument."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.4673108895143003"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N = 1000\n",
    "mean_T_mrca = 0\n",
    "for ts in msprime.simulate(10, num_replicates=N):\n",
    "    tree = ts.first()\n",
    "    mean_T_mrca += tree.time(tree.root)\n",
    "mean_T_mrca = mean_T_mrca / N\n",
    "mean_T_mrca"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example we run 1000 independent replicates of the coalescent for a sample of 10 chromosomes, and compute the mean time to the most recent common ancestor of the entire sample, or equivalently the root of the tree. (The value we obtain is a highly unrealistic 3.5 generations in the past; however, this is unsurprising since we have specified a highly unrealistic population model. See the next for more details on how to specify population models and interpret simulation times.) The ``simulate`` function behaves slightly differently when it is called with the ``num_replicates`` argument: rather than returning a single tree sequence we return an *iterator* over the individual replicates. What this essentially means is that we can use the convenient **for** loop construction to consider each simulation in turn, but without actually storing all these simulations. As a result, we can run millions of replicates using this method if needs be without using any extra storage.\n",
    "\n",
    "When simulating coalescent trees, we are often interested in more than just the mean of the distribution of some statistic. Rather than compute the various summaries by hand (as we have done for the mean in the last example), it is much more convenient to store the result for each replicate and analyse this array after the simulations have completed. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3.4746305209653428, 3.9000652640189766)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N = 1000\n",
    "T_mrca = np.zeros(N)\n",
    "for j, ts in enumerate(msprime.simulate(10, num_replicates=N)):\n",
    "    tree = ts.first()\n",
    "    T_mrca[j] = tree.time(tree.root)\n",
    "np.mean(T_mrca), np.var(T_mrca)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we simulate 1000 replicates, storing the time to the most recent common ancestor for each replicate in the array ``T_mrca``. (By default, time is measured in units of $4 N_e$ generations; see the next section for details.) We use the Python ``enumerate`` function here to simplify the process of efficiently inserting values into this array, which simply ensures that ``j`` for the first replicate, ``1`` for the second, and so on. Thus, by the time we finish the loop, the array has been filled with T_mrca values randomly generated under the coalescent. We then use the ``numpy`` library (which has an extensive suite of statistical functions) to compute the mean and variance of this array. This example is idiomatic, and we will use this type of approach throughout the rest of the tutorial.\n",
    "\n",
    "It is usually more convenient to use the ``num_replicates`` argument to peform replication, but there are situations in which it is better the specify random seeds manually. For example, if simulations require a long time to run then we may wish to use multiple processes to run these simulations. To ensure that the seeds used in these different processes are unique, it is best to manually specify them. For example,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.6002310546233973"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def run_simulation(seed):\n",
    "    # TODO remove the int() here; fixed in version 0.5.0b2.\n",
    "    ts = msprime.simulate(10, random_seed=int(seed))\n",
    "    tree = ts.first()\n",
    "    return tree.time(tree.root)\n",
    "\n",
    "N = 1000\n",
    "seeds = np.random.randint(1, 2**32 - 1, N)\n",
    "with multiprocessing.Pool(4) as pool:\n",
    "    T_mrca = np.array(pool.map(run_simulation, seeds))\n",
    "np.mean(T_mrca)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example we create a list of 1000 seeds between 1 and 2^32 - 1(the range accepted by msprime) randomly using numpy. We then use the multiprocessing module to create a worker pool of four processes, and run our different replicates in these subprocesses. The results are then collected together in a numpy array so that we can easily process them. This approach is a straightforward way to fully utilise modern multi-core processors.\n",
    "\n",
    "Specifying the same random seed for two different simulations (with the same parameters) ensures that we get precisely the same results from both simulations (at least, on the same computer and with the same software versions). This is very useful when we wish to examine the properties of a specific simulation (for example, when debugging), or if we wish to illustrate a particular example. We will often set the random seed in the examples in this tutorial for this reason."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Population models\n",
    "\n",
    "In the previous section the only parameters that we supplied to ``simulate`` were the ``sample_size`` and ``num_replicates`` parameters. This allows us to randomly sample trees with a given number of nodes, but has little connection with biological reality. The most imporant parameter used to model biology in coalescent simulations is the *effective population size*, or ``Ne``. This parameter simply rescales time; larger effective population sizes mean larger coalescence times:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1.016912501990608, 19.770833132209422, 201.84901855758517)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def pairwise_T_mrca(Ne):\n",
    "    N = 10000\n",
    "    T_mrca = np.zeros(N)\n",
    "    for j, ts in enumerate(msprime.simulate(2, Ne=Ne, num_replicates=N)):\n",
    "        tree = ts.first()\n",
    "        T_mrca[j] = tree.time(tree.root)\n",
    "    return np.mean(T_mrca)\n",
    "\n",
    "pairwise_T_mrca(0.5), pairwise_T_mrca(10), pairwise_T_mrca(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Thus, when we specify $N_e=10$ we get a mean coalescence time of about 20 generations, and when $N_e=100$ the mean coalescence time is about 200 generations. See Wakeley (citation) for details on the biological interpretation of effective population size. \n",
    "\n",
    "By default, $N_e = 1$ in msprime, which is equivalent to measuring time in units of $4N_e$ generations. It is very important to note that $N_e$ in msprime is the *diploid* effective population size, which means that all times are scaled by $4N_e$ (rather than 2Ne for a haploid coalescent). Thus, if we wish to compare the results that are given in the literature for a haploid coalescent, then we must set $N_e$ to 1/2 to compensate. For example, we know that the expected coalescence time for a sample of size 2 is 1, and this is the value we obtain from the ``pairwise_T_mrca`` function when we have $N_e=0.5$. We will usually assume that we are working in haploid coalescent time units from here on, and so set $N_e=0.5$ in most examples. However, when running simulations of a specific organism, it is substantially more convenient to use an appropriate estimated value for $N_e$ so that times are directly interpretable.\n",
    "\n",
    "\n",
    "### Exponentially growing/shrinking populations\n",
    "\n",
    "When we provide an $N_e$ parameter, this specifies a fixed effective population size. We can also model populations that are exponentially growing or contracting at some rate over time. Given a population size at the present $s$ and a growth rate $\\alpha$, the size of the population $t$ generations in the past $s e^{-\\alpha t}$. (Note again that time and rates are measured in units of *generations*, not coalescent units.)\n",
    "\n",
    "In msprime, the initial size and growth rate for a particular population are specified using the ``PopulationConfiguration`` object. A list of these objects (describing the different populations; see the Population Structure section) are then provided to the simulate function. When providing a list of ``PopulationConfiguration objects``, the ``Ne`` parameter to ``simulate`` is not required, as the ``initial_size`` of the population configurations performs the same task. For example,\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.91538027853454373, 0.99681829214938678, 1.1158316008471061)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def pairwise_T_mrca(growth_rate):\n",
    "    N = 10000\n",
    "    T_mrca = np.zeros(N)    \n",
    "    replicates = msprime.simulate(\n",
    "        population_configurations=[\n",
    "             msprime.PopulationConfiguration(\n",
    "                sample_size=2, initial_size=0.5, growth_rate=growth_rate)], \n",
    "        num_replicates=N)\n",
    "    for j, ts in enumerate(replicates):\n",
    "        tree = ts.first()\n",
    "        T_mrca[j] = tree.time(tree.root)\n",
    "    return np.mean(T_mrca)\n",
    "\n",
    "pairwise_T_mrca(0.1), pairwise_T_mrca(0), pairwise_T_mrca(-0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we simulate the pairwise $T_{MRCA}$ for positive, zero and negative growth rates. When we have a growth rate of zero, we see that recover the usual result of 1.0 (as our initial size, and hence $N_e$, is set to $1/2$). When the growth rate is positive, we see that the mean coalescence time is reduced, since the population size is getting smaller as we go backwards in time, resulting in an increased rate of coalescence. Conversely, when we have a negative growth rate, the population is getting larger as we go backwards in time, resulting in a slower coalescence rate."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mutations\n",
    "\n",
    "We cannot directly observe gene genealogies; rather we observe the effects of mutations occuring on the branches of these trees. We are there very often interesting not just in the genealogies that are simulated by the coalescent process, but also in the results of mutational processes imposed on these trees. Msprime currently supports simulating mutations under the infinitely many sites model. This is accessed by the ``mutation_rate`` parameter to the ``simulate`` function. As usual, this rate is the per-generation rate.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs/><g id=\"lines\" stroke=\"black\"><line x1=\"40.0\" x2=\"40.0\" y1=\"180.0\" y2=\"20.0\"/><line x1=\"40.0\" x2=\"70.0\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"100.0\" x2=\"100.0\" y1=\"169.6899167243002\" y2=\"20.0\"/><line x1=\"100.0\" x2=\"70.0\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"80.0\" x2=\"80.0\" y1=\"180.0\" y2=\"169.6899167243002\"/><line x1=\"80.0\" x2=\"100.0\" y1=\"169.6899167243002\" y2=\"169.6899167243002\"/><line x1=\"120.0\" x2=\"120.0\" y1=\"180.0\" y2=\"169.6899167243002\"/><line x1=\"120.0\" x2=\"100.0\" y1=\"169.6899167243002\" y2=\"169.6899167243002\"/></g><g font-size=\"14\" text-anchor=\"start\"><text x=\"105.0\" y=\"164.6899167243002\">3</text></g><g font-size=\"14\" text-anchor=\"end\"/><g font-size=\"14\" text-anchor=\"middle\"><text x=\"70.0\" y=\"15.0\">4</text><text x=\"40.0\" y=\"200.0\">1</text><text x=\"80.0\" y=\"200.0\">0</text><text x=\"120.0\" y=\"200.0\">2</text></g><circle cx=\"70.0\" cy=\"20.0\" fill=\"black\" r=\"3\"/><circle cx=\"40.0\" cy=\"180.0\" fill=\"black\" r=\"3\"/><circle cx=\"100.0\" cy=\"169.6899167243002\" fill=\"black\" r=\"3\"/><circle cx=\"80.0\" cy=\"180.0\" fill=\"black\" r=\"3\"/><circle cx=\"120.0\" cy=\"180.0\" fill=\"black\" r=\"3\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"start\"><text x=\"105.0\" y=\"99.3449583621501\">0</text></g><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"end\"><text x=\"35.0\" y=\"104.5\">1</text></g><rect fill=\"red\" height=\"6\" width=\"6\" x=\"37.0\" y=\"97.0\"/><rect fill=\"red\" height=\"6\" width=\"6\" x=\"97.0\" y=\"91.8449583621501\"/></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.simulate(3, mutation_rate=1, random_seed=7)\n",
    "tree = ts.first()\n",
    "SVG(tree.draw())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this example we have two mutations, shown by the red squares. Mutations occur above a given node in the tree, and all samples beneath this node will inherit the mutation. The infinite sites mutations used here are simple binary mutations, so that the ancestral state is 0 and the derived state is 1. One convenient way to access the resulting sample genotypes is to use the ``genotype_matrix()`` method, which returns an $m \\times n$ numpy array, if we have $m$ variable sites and $n$ samples. Thus, if $G$ is the genotype matrix, $G[j, k]$ is the state of the $k$th sample at the $j$th site. In our example above, the site 0 has a mutation over node 3, and site 1 has a mutation over node 1, and so we get the following matrix:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1, 0, 1],\n",
       "       [0, 1, 0]], dtype=uint8)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts.genotype_matrix()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The genotype matrix is provides a convenient way of accessing this information, but will consume a great deal of memory for larger simulations. Msprime has an extensive suite of tools for efficiently working with sequence data and computing population genetics statistics, and a full review of these tools is beyond the scope of this document. Please consult the online documentation for more information on processing simulated sequences.\n",
    "\n",
    "When comparing with analytical results regarding neutral genetic variation, it is very imporant to be aware of the way in which the mutation rates are defined in coalescent theory. For historical reasons, the scaled mutation rate $\\theta$ is defined as $2N_e \\mu$, where $\\mu$ is the per-generation mutation rate. Since all times and rates are specified in units of generations in msprime, we must divide this factor of two out if we are to compare with analytical predictions. For example, the mean number of segregating sites for a sample of two is $\\theta$; to run this in msprime we do the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.9523999999999999"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "N = 10000\n",
    "theta = 5\n",
    "S = np.zeros(N)\n",
    "replicates = msprime.simulate(2, Ne=0.5, mutation_rate=theta / 2, num_replicates=N)\n",
    "for j, ts in enumerate(replicates):\n",
    "    S[j] = ts.num_sites  # Number of segregrating sites.\n",
    "np.mean(S)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that here we set the mutation rate to $\\theta / 2$ (to cancel out the factor of 2 in the definition of $\\theta$) and $N_e = 1/2$ (so that time is measured in haploid coalescent time units). Such factor-of-two gymnastics are unfortunately unavoidable in coalescent theory."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Population structure\n",
    "\n",
    "Msprime supports a discrete-deme model of population structure in which $d$ panmictic populations exchange migrants according to the rates defined in an $d \\times d$ matrix. This approach is very flexible, allowing us to simulate island models (in which all populations exchange migrants at a fixed rate), one and two dimensional stepping stone models (where migrants only move to adjacent demes) and other more complex migration patterns.\n",
    "\n",
    "This population structure is declared in msprime using the ``population_configurations`` and ``migration_matrix`` parameters to the ``simulate`` function. The list of population configurations defines the populations; each element of this list must be a ``PopulationConfiguration`` instance (each population has independent initial population size and growth rate parameters). The migration matrix is a numpy array (or list of lists) of per-generation migration rates; $M[j,k]$ defines the fraction of population $j$ that consists of migrants from population $k$ in each generation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs/><g id=\"lines\" stroke=\"black\"><line x1=\"33.333333333333336\" x2=\"33.333333333333336\" y1=\"180.0\" y2=\"20.0\"/><line x1=\"33.333333333333336\" x2=\"62.5\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"91.66666666666667\" x2=\"91.66666666666667\" y1=\"133.03317526011494\" y2=\"20.0\"/><line x1=\"91.66666666666667\" x2=\"62.5\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"66.66666666666667\" x2=\"66.66666666666667\" y1=\"180.0\" y2=\"133.03317526011494\"/><line x1=\"66.66666666666667\" x2=\"91.66666666666667\" y1=\"133.03317526011494\" y2=\"133.03317526011494\"/><line x1=\"116.66666666666667\" x2=\"116.66666666666667\" y1=\"162.2410632841496\" y2=\"133.03317526011494\"/><line x1=\"116.66666666666667\" x2=\"91.66666666666667\" y1=\"133.03317526011494\" y2=\"133.03317526011494\"/><line x1=\"100.0\" x2=\"100.0\" y1=\"180.0\" y2=\"162.2410632841496\"/><line x1=\"100.0\" x2=\"116.66666666666667\" y1=\"162.2410632841496\" y2=\"162.2410632841496\"/><line x1=\"133.33333333333334\" x2=\"133.33333333333334\" y1=\"180.0\" y2=\"162.2410632841496\"/><line x1=\"133.33333333333334\" x2=\"116.66666666666667\" y1=\"162.2410632841496\" y2=\"162.2410632841496\"/></g><g font-size=\"14\" text-anchor=\"start\"><text x=\"96.66666666666667\" y=\"128.03317526011494\">5</text><text x=\"121.66666666666667\" y=\"157.2410632841496\">4</text></g><g font-size=\"14\" text-anchor=\"end\"/><g font-size=\"14\" text-anchor=\"middle\"><text x=\"62.5\" y=\"15.0\">6</text><text x=\"33.333333333333336\" y=\"200.0\">1</text><text x=\"66.66666666666667\" y=\"200.0\">0</text><text x=\"100.0\" y=\"200.0\">2</text><text x=\"133.33333333333334\" y=\"200.0\">3</text></g><circle cx=\"62.5\" cy=\"20.0\" fill=\"blue\" r=\"3\"/><circle cx=\"33.333333333333336\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"91.66666666666667\" cy=\"133.03317526011494\" fill=\"blue\" r=\"3\"/><circle cx=\"66.66666666666667\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"116.66666666666667\" cy=\"162.2410632841496\" fill=\"blue\" r=\"3\"/><circle cx=\"100.0\" cy=\"180.0\" fill=\"blue\" r=\"3\"/><circle cx=\"133.33333333333334\" cy=\"180.0\" fill=\"blue\" r=\"3\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"start\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"end\"/></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop_configs = [\n",
    "    msprime.PopulationConfiguration(sample_size=2),\n",
    "    msprime.PopulationConfiguration(sample_size=2)\n",
    "]\n",
    "M = np.array([\n",
    "    [0, 0.1],\n",
    "    [0, 0]\n",
    "])\n",
    "ts  = msprime.simulate(\n",
    "    population_configurations=pop_configs, migration_matrix=M,\n",
    "    random_seed=2)\n",
    "tree = ts.first()\n",
    "colour_map = {0:\"red\", 1:\"blue\"}\n",
    "node_colours = {u: colour_map[tree.population(u)] for u in tree.nodes()}\n",
    "SVG(tree.draw(node_colours=node_colours))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We create our population model by first making a list of two ``PopulationConfiguration`` objects. For convenience here, we use the ``sample_size`` argument to these objects to state that we wish to have two samples from each population. This results in samples being allocated sequentially to the populations when ``simulate`` is called: 0 and 1 are placed in population 0, and samples 2 and 3 are place in population 1. We then declare our migration matrix, which is asymmetric in this example. Because ``M[0,1] = 0.1`` and ``M[0,1] = 0``, forwards in time, individuals can migrate from population 1 to population 0 but not vice-versa. This is shown in the tree for this simulation, where each node has been coloured by its population (red is population 0 and blue population 1). Thus, the leaf nodes 0 and 1 are both from population 0, and 2 and 3 are both from population 2 (as explained above). As we go up the tree, the first event that occurs is 2 and 3 coalescing in population 1, creating node 4. After this, 4 coalesces with with node 0, which has at some point before this migrated into deme 1, creating node 5. Node 1 also migrates into deme 1, where it coalesces with 5. Because migration is asymmetric here, the most recent common ancestor of the four samples *must* occur within deme 1.\n",
    "\n",
    "The exact history of migration events is available if we use the ``record_migrations`` option. In the next example, we set up a symmetric island model and track every migration event:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs/><g id=\"lines\" stroke=\"black\"><line x1=\"28.571428571428573\" x2=\"28.571428571428573\" y1=\"180.0\" y2=\"20.0\"/><line x1=\"28.571428571428573\" x2=\"55.35714285714286\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"82.14285714285715\" x2=\"82.14285714285715\" y1=\"141.46049023690188\" y2=\"20.0\"/><line x1=\"82.14285714285715\" x2=\"55.35714285714286\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"57.142857142857146\" x2=\"57.142857142857146\" y1=\"180.0\" y2=\"141.46049023690188\"/><line x1=\"57.142857142857146\" x2=\"82.14285714285715\" y1=\"141.46049023690188\" y2=\"141.46049023690188\"/><line x1=\"107.14285714285715\" x2=\"107.14285714285715\" y1=\"146.4896894979181\" y2=\"141.46049023690188\"/><line x1=\"107.14285714285715\" x2=\"82.14285714285715\" y1=\"141.46049023690188\" y2=\"141.46049023690188\"/><line x1=\"85.71428571428572\" x2=\"85.71428571428572\" y1=\"180.0\" y2=\"146.4896894979181\"/><line x1=\"85.71428571428572\" x2=\"107.14285714285715\" y1=\"146.4896894979181\" y2=\"146.4896894979181\"/><line x1=\"128.57142857142858\" x2=\"128.57142857142858\" y1=\"173.1998053527371\" y2=\"146.4896894979181\"/><line x1=\"128.57142857142858\" x2=\"107.14285714285715\" y1=\"146.4896894979181\" y2=\"146.4896894979181\"/><line x1=\"114.28571428571429\" x2=\"114.28571428571429\" y1=\"180.0\" y2=\"173.1998053527371\"/><line x1=\"114.28571428571429\" x2=\"128.57142857142858\" y1=\"173.1998053527371\" y2=\"173.1998053527371\"/><line x1=\"142.85714285714286\" x2=\"142.85714285714286\" y1=\"180.0\" y2=\"173.1998053527371\"/><line x1=\"142.85714285714286\" x2=\"128.57142857142858\" y1=\"173.1998053527371\" y2=\"173.1998053527371\"/></g><g font-size=\"14\" text-anchor=\"start\"><text x=\"87.14285714285715\" y=\"136.46049023690188\">7</text><text x=\"112.14285714285715\" y=\"141.4896894979181\">6</text><text x=\"133.57142857142858\" y=\"168.1998053527371\">5</text></g><g font-size=\"14\" text-anchor=\"end\"/><g font-size=\"14\" text-anchor=\"middle\"><text x=\"55.35714285714286\" y=\"15.0\">8</text><text x=\"28.571428571428573\" y=\"200.0\">2</text><text x=\"57.142857142857146\" y=\"200.0\">1</text><text x=\"85.71428571428572\" y=\"200.0\">3</text><text x=\"114.28571428571429\" y=\"200.0\">0</text><text x=\"142.85714285714286\" y=\"200.0\">4</text></g><circle cx=\"55.35714285714286\" cy=\"20.0\" fill=\"red\" r=\"3\"/><circle cx=\"28.571428571428573\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"82.14285714285715\" cy=\"141.46049023690188\" fill=\"red\" r=\"3\"/><circle cx=\"57.142857142857146\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"107.14285714285715\" cy=\"146.4896894979181\" fill=\"green\" r=\"3\"/><circle cx=\"85.71428571428572\" cy=\"180.0\" fill=\"blue\" r=\"3\"/><circle cx=\"128.57142857142858\" cy=\"173.1998053527371\" fill=\"green\" r=\"3\"/><circle cx=\"114.28571428571429\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"142.85714285714286\" cy=\"180.0\" fill=\"green\" r=\"3\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"start\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"end\"/></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop_configs = [\n",
    "    msprime.PopulationConfiguration(sample_size=3),\n",
    "    msprime.PopulationConfiguration(sample_size=1),\n",
    "    msprime.PopulationConfiguration(sample_size=1)]\n",
    "M = [\n",
    "    [0, 1, 1],\n",
    "    [1, 0, 1],\n",
    "    [1, 1, 0]]\n",
    "ts  = msprime.simulate(\n",
    "    population_configurations=pop_configs, migration_matrix=M, record_migrations=True,\n",
    "    random_seed=101)\n",
    "tree = ts.first()\n",
    "colour_map = {0:\"red\", 1:\"blue\", 2: \"green\"}\n",
    "node_colours = {u: colour_map[tree.population(u)] for u in tree.nodes()}\n",
    "SVG(tree.draw(node_colours=node_colours))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we sample three nodes from population 0, but because we have a lot of migration the locations of coalescences are quite random. For example, the first coalescence occurs in deme 2 (green), after node zero has migrated in. To see the details of these migration events, we can examine the \"migration records\" that are stored by msprime. (These are not stored by default, as they may require a substantial amount of memory to store. The ``record_migrations`` parameter must be supplied to ``simulate`` to turn on this feature.) Migration records store complete information about the time, source and destination demes and the genomic interval in question. Here we are interested in the total number of migrations experienced by each node:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Created with matplotlib (http://matplotlib.org/) -->\n",
       "<svg height=\"262pt\" version=\"1.1\" viewBox=\"0 0 381 262\" width=\"381pt\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       " <defs>\n",
       "  <style type=\"text/css\">\n",
       "*{stroke-linecap:butt;stroke-linejoin:round;}\n",
       "  </style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M -0 262.415937 \n",
       "L 381.575937 262.415937 \n",
       "L 381.575937 0 \n",
       "L -0 0 \n",
       "z\n",
       "\" style=\"fill:none;\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 39.575938 224.64 \n",
       "L 374.375938 224.64 \n",
       "L 374.375938 7.2 \n",
       "L 39.575938 7.2 \n",
       "z\n",
       "\" style=\"fill:#eaeaf2;\"/>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\">\n",
       "     <g id=\"line2d_1\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 68.62883 224.64 \n",
       "L 68.62883 7.2 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_2\"/>\n",
       "     <g id=\"text_1\">\n",
       "      <!-- 0 -->\n",
       "      <defs>\n",
       "       <path d=\"M 51.703125 34.421875 \n",
       "Q 51.703125 24.515625 49.828125 17.75 \n",
       "Q 47.953125 10.984375 44.703125 6.8125 \n",
       "Q 41.453125 2.640625 37.0625 0.828125 \n",
       "Q 32.671875 -0.984375 27.6875 -0.984375 \n",
       "Q 22.65625 -0.984375 18.3125 0.828125 \n",
       "Q 13.96875 2.640625 10.765625 6.78125 \n",
       "Q 7.5625 10.9375 5.734375 17.703125 \n",
       "Q 3.90625 24.46875 3.90625 34.421875 \n",
       "Q 3.90625 44.828125 5.734375 51.640625 \n",
       "Q 7.5625 58.453125 10.78125 62.5 \n",
       "Q 14.015625 66.546875 18.40625 68.1875 \n",
       "Q 22.796875 69.828125 27.984375 69.828125 \n",
       "Q 32.90625 69.828125 37.21875 68.1875 \n",
       "Q 41.546875 66.546875 44.765625 62.5 \n",
       "Q 48 58.453125 49.84375 51.640625 \n",
       "Q 51.703125 44.828125 51.703125 34.421875 \n",
       "z\n",
       "M 42.78125 34.421875 \n",
       "Q 42.78125 42.625 41.796875 48.0625 \n",
       "Q 40.828125 53.515625 38.921875 56.765625 \n",
       "Q 37.015625 60.015625 34.25 61.359375 \n",
       "Q 31.5 62.703125 27.984375 62.703125 \n",
       "Q 24.265625 62.703125 21.4375 61.328125 \n",
       "Q 18.609375 59.96875 16.671875 56.71875 \n",
       "Q 14.75 53.46875 13.765625 48.015625 \n",
       "Q 12.796875 42.578125 12.796875 34.421875 \n",
       "Q 12.796875 26.515625 13.796875 21.09375 \n",
       "Q 14.796875 15.671875 16.71875 12.375 \n",
       "Q 18.65625 9.078125 21.4375 7.640625 \n",
       "Q 24.21875 6.203125 27.78125 6.203125 \n",
       "Q 31.25 6.203125 34.03125 7.640625 \n",
       "Q 36.8125 9.078125 38.734375 12.375 \n",
       "Q 40.671875 15.671875 41.71875 21.09375 \n",
       "Q 42.78125 26.515625 42.78125 34.421875 \n",
       "z\n",
       "\" id=\"LiberationSans-30\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(65.848361 238.886875)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_2\">\n",
       "     <g id=\"line2d_3\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 137.802384 224.64 \n",
       "L 137.802384 7.2 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_4\"/>\n",
       "     <g id=\"text_2\">\n",
       "      <!-- 2 -->\n",
       "      <defs>\n",
       "       <path d=\"M 5.03125 0 \n",
       "L 5.03125 6.203125 \n",
       "Q 7.515625 11.921875 11.109375 16.28125 \n",
       "Q 14.703125 20.65625 18.65625 24.1875 \n",
       "Q 22.609375 27.734375 26.484375 30.765625 \n",
       "Q 30.375 33.796875 33.5 36.8125 \n",
       "Q 36.625 39.84375 38.546875 43.15625 \n",
       "Q 40.484375 46.484375 40.484375 50.6875 \n",
       "Q 40.484375 53.609375 39.59375 55.828125 \n",
       "Q 38.71875 58.0625 37.0625 59.5625 \n",
       "Q 35.40625 61.078125 33.078125 61.828125 \n",
       "Q 30.765625 62.59375 27.9375 62.59375 \n",
       "Q 25.296875 62.59375 22.96875 61.859375 \n",
       "Q 20.65625 61.140625 18.84375 59.671875 \n",
       "Q 17.046875 58.203125 15.890625 56.03125 \n",
       "Q 14.75 53.859375 14.40625 50.984375 \n",
       "L 5.421875 51.8125 \n",
       "Q 5.859375 55.515625 7.46875 58.78125 \n",
       "Q 9.078125 62.0625 11.90625 64.53125 \n",
       "Q 14.75 67 18.71875 68.40625 \n",
       "Q 22.703125 69.828125 27.9375 69.828125 \n",
       "Q 33.0625 69.828125 37.0625 68.609375 \n",
       "Q 41.0625 67.390625 43.8125 64.984375 \n",
       "Q 46.578125 62.59375 48.046875 59.078125 \n",
       "Q 49.515625 55.5625 49.515625 50.984375 \n",
       "Q 49.515625 47.515625 48.265625 44.390625 \n",
       "Q 47.015625 41.265625 44.9375 38.421875 \n",
       "Q 42.875 35.59375 40.140625 32.953125 \n",
       "Q 37.40625 30.328125 34.421875 27.8125 \n",
       "Q 31.453125 25.296875 28.421875 22.828125 \n",
       "Q 25.390625 20.359375 22.71875 17.859375 \n",
       "Q 20.0625 15.375 17.96875 12.8125 \n",
       "Q 15.875 10.25 14.703125 7.46875 \n",
       "L 50.59375 7.46875 \n",
       "L 50.59375 0 \n",
       "z\n",
       "\" id=\"LiberationSans-32\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(135.021915 238.886875)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-32\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_3\">\n",
       "     <g id=\"line2d_5\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 206.975938 224.64 \n",
       "L 206.975938 7.2 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_6\"/>\n",
       "     <g id=\"text_3\">\n",
       "      <!-- 4 -->\n",
       "      <defs>\n",
       "       <path d=\"M 43.015625 15.578125 \n",
       "L 43.015625 0 \n",
       "L 34.71875 0 \n",
       "L 34.71875 15.578125 \n",
       "L 2.296875 15.578125 \n",
       "L 2.296875 22.40625 \n",
       "L 33.796875 68.796875 \n",
       "L 43.015625 68.796875 \n",
       "L 43.015625 22.515625 \n",
       "L 52.6875 22.515625 \n",
       "L 52.6875 15.578125 \n",
       "z\n",
       "M 34.71875 58.890625 \n",
       "Q 34.625 58.640625 34.234375 57.9375 \n",
       "Q 33.84375 57.234375 33.34375 56.34375 \n",
       "Q 32.859375 55.46875 32.34375 54.5625 \n",
       "Q 31.84375 53.65625 31.453125 53.078125 \n",
       "L 13.8125 27.09375 \n",
       "Q 13.578125 26.703125 13.109375 26.0625 \n",
       "Q 12.640625 25.4375 12.15625 24.78125 \n",
       "Q 11.671875 24.125 11.171875 23.484375 \n",
       "Q 10.6875 22.859375 10.40625 22.515625 \n",
       "L 34.71875 22.515625 \n",
       "z\n",
       "\" id=\"LiberationSans-34\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(204.195469 238.886875)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-34\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_4\">\n",
       "     <g id=\"line2d_7\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 276.149491 224.64 \n",
       "L 276.149491 7.2 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_8\"/>\n",
       "     <g id=\"text_4\">\n",
       "      <!-- 6 -->\n",
       "      <defs>\n",
       "       <path d=\"M 51.21875 22.515625 \n",
       "Q 51.21875 17.328125 49.78125 13 \n",
       "Q 48.34375 8.6875 45.53125 5.578125 \n",
       "Q 42.71875 2.484375 38.5625 0.75 \n",
       "Q 34.421875 -0.984375 29 -0.984375 \n",
       "Q 23 -0.984375 18.5 1.3125 \n",
       "Q 14.015625 3.609375 11.03125 7.921875 \n",
       "Q 8.0625 12.25 6.5625 18.53125 \n",
       "Q 5.078125 24.8125 5.078125 32.8125 \n",
       "Q 5.078125 42 6.765625 48.921875 \n",
       "Q 8.453125 55.859375 11.625 60.5 \n",
       "Q 14.796875 65.140625 19.359375 67.484375 \n",
       "Q 23.921875 69.828125 29.6875 69.828125 \n",
       "Q 33.203125 69.828125 36.28125 69.09375 \n",
       "Q 39.359375 68.359375 41.875 66.71875 \n",
       "Q 44.390625 65.09375 46.28125 62.40625 \n",
       "Q 48.1875 59.71875 49.3125 55.8125 \n",
       "L 40.921875 54.296875 \n",
       "Q 39.546875 58.734375 36.546875 60.71875 \n",
       "Q 33.546875 62.703125 29.59375 62.703125 \n",
       "Q 25.984375 62.703125 23.046875 60.984375 \n",
       "Q 20.125 59.28125 18.0625 55.875 \n",
       "Q 16.015625 52.484375 14.90625 47.359375 \n",
       "Q 13.8125 42.234375 13.8125 35.40625 \n",
       "Q 16.21875 39.84375 20.5625 42.15625 \n",
       "Q 24.90625 44.484375 30.515625 44.484375 \n",
       "Q 35.203125 44.484375 39.015625 42.96875 \n",
       "Q 42.828125 41.453125 45.53125 38.59375 \n",
       "Q 48.25 35.75 49.734375 31.671875 \n",
       "Q 51.21875 27.59375 51.21875 22.515625 \n",
       "z\n",
       "M 42.28125 22.125 \n",
       "Q 42.28125 25.6875 41.40625 28.5625 \n",
       "Q 40.53125 31.453125 38.765625 33.46875 \n",
       "Q 37.015625 35.5 34.421875 36.59375 \n",
       "Q 31.84375 37.703125 28.421875 37.703125 \n",
       "Q 26.03125 37.703125 23.578125 36.984375 \n",
       "Q 21.140625 36.28125 19.15625 34.6875 \n",
       "Q 17.1875 33.109375 15.9375 30.515625 \n",
       "Q 14.703125 27.9375 14.703125 24.21875 \n",
       "Q 14.703125 20.40625 15.671875 17.109375 \n",
       "Q 16.65625 13.8125 18.484375 11.375 \n",
       "Q 20.3125 8.9375 22.890625 7.515625 \n",
       "Q 25.484375 6.109375 28.71875 6.109375 \n",
       "Q 31.890625 6.109375 34.40625 7.203125 \n",
       "Q 36.921875 8.296875 38.671875 10.375 \n",
       "Q 40.4375 12.453125 41.359375 15.421875 \n",
       "Q 42.28125 18.40625 42.28125 22.125 \n",
       "z\n",
       "\" id=\"LiberationSans-36\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(273.369022 238.886875)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-36\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_5\">\n",
       "     <g id=\"line2d_9\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 345.323045 224.64 \n",
       "L 345.323045 7.2 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_10\"/>\n",
       "     <g id=\"text_5\">\n",
       "      <!-- 8 -->\n",
       "      <defs>\n",
       "       <path d=\"M 51.265625 19.1875 \n",
       "Q 51.265625 14.796875 49.875 11.109375 \n",
       "Q 48.484375 7.421875 45.625 4.734375 \n",
       "Q 42.78125 2.046875 38.328125 0.53125 \n",
       "Q 33.890625 -0.984375 27.828125 -0.984375 \n",
       "Q 21.78125 -0.984375 17.359375 0.53125 \n",
       "Q 12.9375 2.046875 10.03125 4.703125 \n",
       "Q 7.125 7.375 5.734375 11.0625 \n",
       "Q 4.34375 14.75 4.34375 19.09375 \n",
       "Q 4.34375 22.859375 5.484375 25.78125 \n",
       "Q 6.640625 28.71875 8.5625 30.828125 \n",
       "Q 10.5 32.953125 12.96875 34.25 \n",
       "Q 15.4375 35.546875 18.0625 35.984375 \n",
       "L 18.0625 36.1875 \n",
       "Q 15.1875 36.859375 12.90625 38.375 \n",
       "Q 10.640625 39.890625 9.09375 42.015625 \n",
       "Q 7.5625 44.140625 6.75 46.71875 \n",
       "Q 5.953125 49.3125 5.953125 52.203125 \n",
       "Q 5.953125 55.8125 7.34375 59 \n",
       "Q 8.734375 62.203125 11.46875 64.625 \n",
       "Q 14.203125 67.046875 18.25 68.4375 \n",
       "Q 22.3125 69.828125 27.640625 69.828125 \n",
       "Q 33.25 69.828125 37.375 68.40625 \n",
       "Q 41.5 67 44.203125 64.578125 \n",
       "Q 46.921875 62.15625 48.234375 58.9375 \n",
       "Q 49.5625 55.71875 49.5625 52.09375 \n",
       "Q 49.5625 49.265625 48.75 46.671875 \n",
       "Q 47.953125 44.09375 46.40625 41.96875 \n",
       "Q 44.875 39.84375 42.59375 38.34375 \n",
       "Q 40.328125 36.859375 37.359375 36.28125 \n",
       "L 37.359375 36.078125 \n",
       "Q 40.328125 35.59375 42.859375 34.296875 \n",
       "Q 45.40625 33.015625 47.265625 30.890625 \n",
       "Q 49.125 28.765625 50.1875 25.828125 \n",
       "Q 51.265625 22.90625 51.265625 19.1875 \n",
       "z\n",
       "M 40.4375 51.609375 \n",
       "Q 40.4375 54.203125 39.765625 56.34375 \n",
       "Q 39.109375 58.5 37.59375 60.03125 \n",
       "Q 36.078125 61.578125 33.640625 62.421875 \n",
       "Q 31.203125 63.28125 27.640625 63.28125 \n",
       "Q 24.171875 63.28125 21.78125 62.421875 \n",
       "Q 19.390625 61.578125 17.84375 60.03125 \n",
       "Q 16.3125 58.5 15.625 56.34375 \n",
       "Q 14.9375 54.203125 14.9375 51.609375 \n",
       "Q 14.9375 49.5625 15.46875 47.40625 \n",
       "Q 16.015625 45.265625 17.421875 43.5 \n",
       "Q 18.84375 41.75 21.328125 40.625 \n",
       "Q 23.828125 39.5 27.734375 39.5 \n",
       "Q 31.890625 39.5 34.40625 40.625 \n",
       "Q 36.921875 41.75 38.25 43.5 \n",
       "Q 39.59375 45.265625 40.015625 47.40625 \n",
       "Q 40.4375 49.5625 40.4375 51.609375 \n",
       "z\n",
       "M 42.140625 20.015625 \n",
       "Q 42.140625 22.515625 41.453125 24.828125 \n",
       "Q 40.765625 27.15625 39.109375 28.9375 \n",
       "Q 37.453125 30.71875 34.640625 31.8125 \n",
       "Q 31.84375 32.90625 27.640625 32.90625 \n",
       "Q 23.78125 32.90625 21.0625 31.8125 \n",
       "Q 18.359375 30.71875 16.671875 28.90625 \n",
       "Q 14.984375 27.09375 14.203125 24.71875 \n",
       "Q 13.421875 22.359375 13.421875 19.828125 \n",
       "Q 13.421875 16.65625 14.203125 14.03125 \n",
       "Q 14.984375 11.421875 16.6875 9.546875 \n",
       "Q 18.40625 7.671875 21.1875 6.640625 \n",
       "Q 23.96875 5.609375 27.9375 5.609375 \n",
       "Q 31.9375 5.609375 34.671875 6.640625 \n",
       "Q 37.40625 7.671875 39.0625 9.546875 \n",
       "Q 40.71875 11.421875 41.421875 14.078125 \n",
       "Q 42.140625 16.75 42.140625 20.015625 \n",
       "z\n",
       "\" id=\"LiberationSans-38\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(342.542576 238.886875)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-38\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_6\">\n",
       "     <!-- Node ID -->\n",
       "     <defs>\n",
       "      <path d=\"M 52.828125 0 \n",
       "L 16.015625 58.59375 \n",
       "Q 16.109375 56.203125 16.265625 53.859375 \n",
       "Q 16.359375 51.859375 16.421875 49.625 \n",
       "Q 16.5 47.40625 16.5 45.703125 \n",
       "L 16.5 0 \n",
       "L 8.203125 0 \n",
       "L 8.203125 68.796875 \n",
       "L 19.046875 68.796875 \n",
       "L 56.25 9.8125 \n",
       "Q 56.109375 12.203125 55.953125 14.59375 \n",
       "Q 55.859375 16.65625 55.765625 19.0625 \n",
       "Q 55.671875 21.484375 55.671875 23.6875 \n",
       "L 55.671875 68.796875 \n",
       "L 64.0625 68.796875 \n",
       "L 64.0625 0 \n",
       "z\n",
       "\" id=\"LiberationSans-4e\"/>\n",
       "      <path d=\"M 51.421875 26.46875 \n",
       "Q 51.421875 12.59375 45.3125 5.796875 \n",
       "Q 39.203125 -0.984375 27.59375 -0.984375 \n",
       "Q 22.078125 -0.984375 17.71875 0.671875 \n",
       "Q 13.375 2.34375 10.375 5.765625 \n",
       "Q 7.375 9.1875 5.78125 14.328125 \n",
       "Q 4.203125 19.484375 4.203125 26.46875 \n",
       "Q 4.203125 53.8125 27.875 53.8125 \n",
       "Q 34.03125 53.8125 38.5 52.09375 \n",
       "Q 42.96875 50.390625 45.828125 46.96875 \n",
       "Q 48.6875 43.5625 50.046875 38.421875 \n",
       "Q 51.421875 33.296875 51.421875 26.46875 \n",
       "z\n",
       "M 42.1875 26.46875 \n",
       "Q 42.1875 32.625 41.234375 36.625 \n",
       "Q 40.28125 40.625 38.453125 43.015625 \n",
       "Q 36.625 45.40625 33.984375 46.359375 \n",
       "Q 31.34375 47.3125 28.03125 47.3125 \n",
       "Q 24.65625 47.3125 21.9375 46.3125 \n",
       "Q 19.234375 45.3125 17.328125 42.890625 \n",
       "Q 15.4375 40.484375 14.421875 36.46875 \n",
       "Q 13.421875 32.46875 13.421875 26.46875 \n",
       "Q 13.421875 20.3125 14.5 16.28125 \n",
       "Q 15.578125 12.25 17.453125 9.859375 \n",
       "Q 19.34375 7.46875 21.90625 6.484375 \n",
       "Q 24.46875 5.515625 27.484375 5.515625 \n",
       "Q 30.859375 5.515625 33.59375 6.46875 \n",
       "Q 36.328125 7.421875 38.234375 9.8125 \n",
       "Q 40.140625 12.203125 41.15625 16.25 \n",
       "Q 42.1875 20.3125 42.1875 26.46875 \n",
       "z\n",
       "\" id=\"LiberationSans-6f\"/>\n",
       "      <path d=\"M 40.09375 8.5 \n",
       "Q 37.640625 3.421875 33.609375 1.21875 \n",
       "Q 29.59375 -0.984375 23.640625 -0.984375 \n",
       "Q 13.625 -0.984375 8.90625 5.75 \n",
       "Q 4.203125 12.5 4.203125 26.171875 \n",
       "Q 4.203125 53.8125 23.640625 53.8125 \n",
       "Q 29.640625 53.8125 33.640625 51.609375 \n",
       "Q 37.640625 49.421875 40.09375 44.625 \n",
       "L 40.1875 44.625 \n",
       "Q 40.1875 45.125 40.15625 46.171875 \n",
       "Q 40.140625 47.21875 40.109375 48.359375 \n",
       "Q 40.09375 49.515625 40.09375 50.53125 \n",
       "Q 40.09375 51.5625 40.09375 52 \n",
       "L 40.09375 72.46875 \n",
       "L 48.875 72.46875 \n",
       "L 48.875 10.890625 \n",
       "Q 48.875 8.984375 48.890625 7.21875 \n",
       "Q 48.921875 5.46875 48.96875 4 \n",
       "Q 49.03125 2.546875 49.078125 1.484375 \n",
       "Q 49.125 0.4375 49.171875 0 \n",
       "L 40.765625 0 \n",
       "Q 40.671875 0.484375 40.59375 1.390625 \n",
       "Q 40.53125 2.296875 40.453125 3.46875 \n",
       "Q 40.375 4.640625 40.328125 5.9375 \n",
       "Q 40.28125 7.234375 40.28125 8.5 \n",
       "z\n",
       "M 13.421875 26.46875 \n",
       "Q 13.421875 21 14.109375 17.09375 \n",
       "Q 14.796875 13.1875 16.3125 10.671875 \n",
       "Q 17.828125 8.15625 20.171875 6.984375 \n",
       "Q 22.515625 5.8125 25.875 5.8125 \n",
       "Q 29.34375 5.8125 32 6.9375 \n",
       "Q 34.671875 8.0625 36.453125 10.578125 \n",
       "Q 38.234375 13.09375 39.15625 17.140625 \n",
       "Q 40.09375 21.1875 40.09375 27.046875 \n",
       "Q 40.09375 32.671875 39.15625 36.546875 \n",
       "Q 38.234375 40.4375 36.421875 42.828125 \n",
       "Q 34.625 45.21875 32 46.265625 \n",
       "Q 29.390625 47.3125 25.984375 47.3125 \n",
       "Q 22.75 47.3125 20.40625 46.1875 \n",
       "Q 18.0625 45.0625 16.5 42.578125 \n",
       "Q 14.9375 40.09375 14.171875 36.109375 \n",
       "Q 13.421875 32.125 13.421875 26.46875 \n",
       "z\n",
       "\" id=\"LiberationSans-64\"/>\n",
       "      <path d=\"M 13.484375 24.5625 \n",
       "Q 13.484375 20.40625 14.328125 16.90625 \n",
       "Q 15.1875 13.421875 16.96875 10.90625 \n",
       "Q 18.75 8.40625 21.53125 7 \n",
       "Q 24.3125 5.609375 28.21875 5.609375 \n",
       "Q 33.9375 5.609375 37.375 7.90625 \n",
       "Q 40.828125 10.203125 42.046875 13.71875 \n",
       "L 49.75 11.53125 \n",
       "Q 48.921875 9.328125 47.4375 7.109375 \n",
       "Q 45.953125 4.890625 43.453125 3.09375 \n",
       "Q 40.96875 1.3125 37.234375 0.15625 \n",
       "Q 33.5 -0.984375 28.21875 -0.984375 \n",
       "Q 16.5 -0.984375 10.375 6 \n",
       "Q 4.25 12.984375 4.25 26.765625 \n",
       "Q 4.25 34.1875 6.09375 39.328125 \n",
       "Q 7.953125 44.484375 11.171875 47.703125 \n",
       "Q 14.40625 50.921875 18.703125 52.359375 \n",
       "Q 23 53.8125 27.875 53.8125 \n",
       "Q 34.515625 53.8125 38.984375 51.65625 \n",
       "Q 43.453125 49.515625 46.15625 45.71875 \n",
       "Q 48.875 41.9375 50.015625 36.8125 \n",
       "Q 51.171875 31.6875 51.171875 25.734375 \n",
       "L 51.171875 24.5625 \n",
       "z\n",
       "M 42.09375 31.296875 \n",
       "Q 41.359375 39.65625 37.84375 43.484375 \n",
       "Q 34.328125 47.3125 27.734375 47.3125 \n",
       "Q 25.53125 47.3125 23.109375 46.609375 \n",
       "Q 20.703125 45.90625 18.65625 44.09375 \n",
       "Q 16.609375 42.28125 15.1875 39.171875 \n",
       "Q 13.765625 36.078125 13.578125 31.296875 \n",
       "z\n",
       "\" id=\"LiberationSans-65\"/>\n",
       "      <path id=\"LiberationSans-20\"/>\n",
       "      <path d=\"M 9.234375 0 \n",
       "L 9.234375 68.796875 \n",
       "L 18.5625 68.796875 \n",
       "L 18.5625 0 \n",
       "z\n",
       "\" id=\"LiberationSans-49\"/>\n",
       "      <path d=\"M 67.4375 35.109375 \n",
       "Q 67.4375 26.515625 64.84375 19.9375 \n",
       "Q 62.25 13.375 57.75 8.953125 \n",
       "Q 53.265625 4.546875 47.140625 2.265625 \n",
       "Q 41.015625 0 33.9375 0 \n",
       "L 8.203125 0 \n",
       "L 8.203125 68.796875 \n",
       "L 30.953125 68.796875 \n",
       "Q 38.921875 68.796875 45.625 66.8125 \n",
       "Q 52.34375 64.84375 57.171875 60.71875 \n",
       "Q 62.015625 56.59375 64.71875 50.234375 \n",
       "Q 67.4375 43.890625 67.4375 35.109375 \n",
       "z\n",
       "M 58.0625 35.109375 \n",
       "Q 58.0625 42.09375 56.046875 47.046875 \n",
       "Q 54.046875 52 50.4375 55.171875 \n",
       "Q 46.828125 58.34375 41.796875 59.828125 \n",
       "Q 36.765625 61.328125 30.765625 61.328125 \n",
       "L 17.53125 61.328125 \n",
       "L 17.53125 7.46875 \n",
       "L 32.859375 7.46875 \n",
       "Q 38.28125 7.46875 42.890625 9.21875 \n",
       "Q 47.515625 10.984375 50.875 14.453125 \n",
       "Q 54.25 17.921875 56.15625 23.09375 \n",
       "Q 58.0625 28.265625 58.0625 35.109375 \n",
       "z\n",
       "\" id=\"LiberationSans-44\"/>\n",
       "     </defs>\n",
       "     <g style=\"fill:#262626;\" transform=\"translate(186.800391 252.933437)scale(0.11 -0.11)\">\n",
       "      <use xlink:href=\"#LiberationSans-4e\"/>\n",
       "      <use x=\"72.216797\" xlink:href=\"#LiberationSans-6f\"/>\n",
       "      <use x=\"127.832031\" xlink:href=\"#LiberationSans-64\"/>\n",
       "      <use x=\"183.447266\" xlink:href=\"#LiberationSans-65\"/>\n",
       "      <use x=\"239.0625\" xlink:href=\"#LiberationSans-20\"/>\n",
       "      <use x=\"266.845703\" xlink:href=\"#LiberationSans-49\"/>\n",
       "      <use x=\"294.628906\" xlink:href=\"#LiberationSans-44\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\">\n",
       "     <g id=\"line2d_11\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 39.575938 224.64 \n",
       "L 374.375938 224.64 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_12\"/>\n",
       "     <g id=\"text_7\">\n",
       "      <!-- 0 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(27.015 228.263437)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_2\">\n",
       "     <g id=\"line2d_13\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 39.575938 190.125714 \n",
       "L 374.375938 190.125714 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_14\"/>\n",
       "     <g id=\"text_8\">\n",
       "      <!-- 5 -->\n",
       "      <defs>\n",
       "       <path d=\"M 51.421875 22.40625 \n",
       "Q 51.421875 17.234375 49.859375 12.9375 \n",
       "Q 48.296875 8.640625 45.21875 5.53125 \n",
       "Q 42.140625 2.4375 37.578125 0.71875 \n",
       "Q 33.015625 -0.984375 27 -0.984375 \n",
       "Q 21.578125 -0.984375 17.546875 0.28125 \n",
       "Q 13.53125 1.5625 10.734375 3.78125 \n",
       "Q 7.953125 6 6.3125 8.984375 \n",
       "Q 4.6875 11.96875 4 15.375 \n",
       "L 12.890625 16.40625 \n",
       "Q 13.421875 14.453125 14.390625 12.625 \n",
       "Q 15.375 10.796875 17.0625 9.34375 \n",
       "Q 18.75 7.90625 21.21875 7.046875 \n",
       "Q 23.6875 6.203125 27.203125 6.203125 \n",
       "Q 30.609375 6.203125 33.390625 7.25 \n",
       "Q 36.1875 8.296875 38.15625 10.34375 \n",
       "Q 40.140625 12.40625 41.203125 15.375 \n",
       "Q 42.28125 18.359375 42.28125 22.21875 \n",
       "Q 42.28125 25.390625 41.25 28.046875 \n",
       "Q 40.234375 30.71875 38.328125 32.640625 \n",
       "Q 36.421875 34.578125 33.65625 35.640625 \n",
       "Q 30.90625 36.71875 27.390625 36.71875 \n",
       "Q 25.203125 36.71875 23.34375 36.328125 \n",
       "Q 21.484375 35.9375 19.890625 35.25 \n",
       "Q 18.3125 34.578125 17.015625 33.671875 \n",
       "Q 15.71875 32.765625 14.59375 31.78125 \n",
       "L 6 31.78125 \n",
       "L 8.296875 68.796875 \n",
       "L 47.40625 68.796875 \n",
       "L 47.40625 61.328125 \n",
       "L 16.3125 61.328125 \n",
       "L 14.984375 39.5 \n",
       "Q 17.328125 41.3125 20.84375 42.59375 \n",
       "Q 24.359375 43.890625 29.203125 43.890625 \n",
       "Q 34.328125 43.890625 38.421875 42.328125 \n",
       "Q 42.53125 40.765625 45.40625 37.90625 \n",
       "Q 48.296875 35.0625 49.859375 31.109375 \n",
       "Q 51.421875 27.15625 51.421875 22.40625 \n",
       "z\n",
       "\" id=\"LiberationSans-35\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(27.015 193.749152)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-35\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_3\">\n",
       "     <g id=\"line2d_15\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 39.575938 155.611429 \n",
       "L 374.375938 155.611429 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_16\"/>\n",
       "     <g id=\"text_9\">\n",
       "      <!-- 10 -->\n",
       "      <defs>\n",
       "       <path d=\"M 7.625 0 \n",
       "L 7.625 7.46875 \n",
       "L 25.140625 7.46875 \n",
       "L 25.140625 60.40625 \n",
       "L 9.625 49.3125 \n",
       "L 9.625 57.625 \n",
       "L 25.875 68.796875 \n",
       "L 33.984375 68.796875 \n",
       "L 33.984375 7.46875 \n",
       "L 50.734375 7.46875 \n",
       "L 50.734375 0 \n",
       "z\n",
       "\" id=\"LiberationSans-31\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(21.454063 159.234866)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-31\"/>\n",
       "       <use x=\"55.615234\" xlink:href=\"#LiberationSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_4\">\n",
       "     <g id=\"line2d_17\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 39.575938 121.097143 \n",
       "L 374.375938 121.097143 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_18\"/>\n",
       "     <g id=\"text_10\">\n",
       "      <!-- 15 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(21.454063 124.72058)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-31\"/>\n",
       "       <use x=\"55.615234\" xlink:href=\"#LiberationSans-35\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_5\">\n",
       "     <g id=\"line2d_19\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 39.575938 86.582857 \n",
       "L 374.375938 86.582857 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_20\"/>\n",
       "     <g id=\"text_11\">\n",
       "      <!-- 20 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(21.454063 90.206295)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-32\"/>\n",
       "       <use x=\"55.615234\" xlink:href=\"#LiberationSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_6\">\n",
       "     <g id=\"line2d_21\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 39.575938 52.068571 \n",
       "L 374.375938 52.068571 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_22\"/>\n",
       "     <g id=\"text_12\">\n",
       "      <!-- 25 -->\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(21.454063 55.692009)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-32\"/>\n",
       "       <use x=\"55.615234\" xlink:href=\"#LiberationSans-35\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_7\">\n",
       "     <g id=\"line2d_23\">\n",
       "      <path clip-path=\"url(#pce3d112b9a)\" d=\"M 39.575938 17.554286 \n",
       "L 374.375938 17.554286 \n",
       "\" style=\"fill:none;stroke:#ffffff;stroke-linecap:round;\"/>\n",
       "     </g>\n",
       "     <g id=\"line2d_24\"/>\n",
       "     <g id=\"text_13\">\n",
       "      <!-- 30 -->\n",
       "      <defs>\n",
       "       <path d=\"M 51.21875 19 \n",
       "Q 51.21875 14.265625 49.671875 10.546875 \n",
       "Q 48.140625 6.84375 45.1875 4.265625 \n",
       "Q 42.234375 1.703125 37.859375 0.359375 \n",
       "Q 33.5 -0.984375 27.875 -0.984375 \n",
       "Q 21.484375 -0.984375 17.109375 0.609375 \n",
       "Q 12.75 2.203125 9.90625 4.8125 \n",
       "Q 7.078125 7.421875 5.65625 10.765625 \n",
       "Q 4.25 14.109375 3.8125 17.671875 \n",
       "L 12.890625 18.5 \n",
       "Q 13.28125 15.765625 14.328125 13.515625 \n",
       "Q 15.375 11.28125 17.1875 9.671875 \n",
       "Q 19 8.0625 21.625 7.171875 \n",
       "Q 24.265625 6.296875 27.875 6.296875 \n",
       "Q 34.515625 6.296875 38.296875 9.5625 \n",
       "Q 42.09375 12.84375 42.09375 19.28125 \n",
       "Q 42.09375 23.09375 40.40625 25.40625 \n",
       "Q 38.71875 27.734375 36.203125 29.03125 \n",
       "Q 33.6875 30.328125 30.734375 30.765625 \n",
       "Q 27.78125 31.203125 25.296875 31.203125 \n",
       "L 20.3125 31.203125 \n",
       "L 20.3125 38.8125 \n",
       "L 25.09375 38.8125 \n",
       "Q 27.59375 38.8125 30.265625 39.328125 \n",
       "Q 32.953125 39.84375 35.171875 41.1875 \n",
       "Q 37.40625 42.53125 38.84375 44.828125 \n",
       "Q 40.28125 47.125 40.28125 50.6875 \n",
       "Q 40.28125 56.203125 37.03125 59.390625 \n",
       "Q 33.796875 62.59375 27.390625 62.59375 \n",
       "Q 21.578125 62.59375 17.984375 59.609375 \n",
       "Q 14.40625 56.640625 13.8125 51.21875 \n",
       "L 4.984375 51.90625 \n",
       "Q 5.515625 56.453125 7.46875 59.8125 \n",
       "Q 9.421875 63.1875 12.421875 65.40625 \n",
       "Q 15.4375 67.625 19.28125 68.71875 \n",
       "Q 23.140625 69.828125 27.484375 69.828125 \n",
       "Q 33.25 69.828125 37.390625 68.375 \n",
       "Q 41.546875 66.9375 44.1875 64.46875 \n",
       "Q 46.828125 62.015625 48.0625 58.6875 \n",
       "Q 49.3125 55.375 49.3125 51.609375 \n",
       "Q 49.3125 48.578125 48.484375 45.9375 \n",
       "Q 47.65625 43.3125 45.890625 41.203125 \n",
       "Q 44.140625 39.109375 41.421875 37.59375 \n",
       "Q 38.71875 36.078125 34.90625 35.296875 \n",
       "L 34.90625 35.109375 \n",
       "Q 39.0625 34.671875 42.140625 33.21875 \n",
       "Q 45.21875 31.78125 47.21875 29.625 \n",
       "Q 49.21875 27.484375 50.21875 24.75 \n",
       "Q 51.21875 22.015625 51.21875 19 \n",
       "z\n",
       "\" id=\"LiberationSans-33\"/>\n",
       "      </defs>\n",
       "      <g style=\"fill:#262626;\" transform=\"translate(21.454063 21.177723)scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#LiberationSans-33\"/>\n",
       "       <use x=\"55.615234\" xlink:href=\"#LiberationSans-30\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_14\">\n",
       "     <!-- Number of migrations -->\n",
       "     <defs>\n",
       "      <path d=\"M 15.328125 52.828125 \n",
       "L 15.328125 19.34375 \n",
       "Q 15.328125 15.484375 15.890625 12.890625 \n",
       "Q 16.453125 10.296875 17.71875 8.703125 \n",
       "Q 19 7.125 21.0625 6.46875 \n",
       "Q 23.140625 5.8125 26.21875 5.8125 \n",
       "Q 29.34375 5.8125 31.859375 6.90625 \n",
       "Q 34.375 8.015625 36.15625 10.078125 \n",
       "Q 37.9375 12.15625 38.90625 15.203125 \n",
       "Q 39.890625 18.265625 39.890625 22.21875 \n",
       "L 39.890625 52.828125 \n",
       "L 48.6875 52.828125 \n",
       "L 48.6875 11.28125 \n",
       "Q 48.6875 9.625 48.703125 7.78125 \n",
       "Q 48.734375 5.953125 48.78125 4.3125 \n",
       "Q 48.828125 2.6875 48.875 1.515625 \n",
       "Q 48.921875 0.34375 48.96875 0 \n",
       "L 40.671875 0 \n",
       "Q 40.625 0.25 40.578125 1.3125 \n",
       "Q 40.53125 2.390625 40.453125 3.78125 \n",
       "Q 40.375 5.171875 40.328125 6.609375 \n",
       "Q 40.28125 8.0625 40.28125 9.03125 \n",
       "L 40.140625 9.03125 \n",
       "Q 38.875 6.734375 37.359375 4.875 \n",
       "Q 35.84375 3.03125 33.84375 1.734375 \n",
       "Q 31.84375 0.4375 29.25 -0.265625 \n",
       "Q 26.65625 -0.984375 23.25 -0.984375 \n",
       "Q 18.84375 -0.984375 15.671875 0.09375 \n",
       "Q 12.5 1.171875 10.453125 3.421875 \n",
       "Q 8.40625 5.671875 7.453125 9.1875 \n",
       "Q 6.5 12.703125 6.5 17.625 \n",
       "L 6.5 52.828125 \n",
       "z\n",
       "\" id=\"LiberationSans-75\"/>\n",
       "      <path d=\"M 37.5 0 \n",
       "L 37.5 33.5 \n",
       "Q 37.5 37.359375 37.015625 39.9375 \n",
       "Q 36.53125 42.53125 35.375 44.109375 \n",
       "Q 34.234375 45.703125 32.375 46.359375 \n",
       "Q 30.515625 47.015625 27.828125 47.015625 \n",
       "Q 25.046875 47.015625 22.796875 45.921875 \n",
       "Q 20.5625 44.828125 18.96875 42.75 \n",
       "Q 17.390625 40.671875 16.53125 37.625 \n",
       "Q 15.671875 34.578125 15.671875 30.609375 \n",
       "L 15.671875 0 \n",
       "L 6.9375 0 \n",
       "L 6.9375 41.546875 \n",
       "Q 6.9375 43.21875 6.90625 45.046875 \n",
       "Q 6.890625 46.875 6.828125 48.5 \n",
       "Q 6.78125 50.140625 6.734375 51.3125 \n",
       "Q 6.6875 52.484375 6.640625 52.828125 \n",
       "L 14.9375 52.828125 \n",
       "Q 14.984375 52.59375 15.03125 51.515625 \n",
       "Q 15.09375 50.4375 15.15625 49.046875 \n",
       "Q 15.234375 47.65625 15.28125 46.21875 \n",
       "Q 15.328125 44.78125 15.328125 43.796875 \n",
       "L 15.484375 43.796875 \n",
       "Q 16.65625 46.09375 18.015625 47.953125 \n",
       "Q 19.390625 49.8125 21.21875 51.09375 \n",
       "Q 23.046875 52.390625 25.40625 53.09375 \n",
       "Q 27.78125 53.8125 30.90625 53.8125 \n",
       "Q 36.921875 53.8125 40.40625 51.421875 \n",
       "Q 43.890625 49.03125 45.265625 43.796875 \n",
       "L 45.40625 43.796875 \n",
       "Q 46.578125 46.09375 48.046875 47.953125 \n",
       "Q 49.515625 49.8125 51.46875 51.09375 \n",
       "Q 53.421875 52.390625 55.859375 53.09375 \n",
       "Q 58.296875 53.8125 61.421875 53.8125 \n",
       "Q 65.4375 53.8125 68.328125 52.734375 \n",
       "Q 71.234375 51.65625 73.09375 49.40625 \n",
       "Q 74.953125 47.171875 75.828125 43.625 \n",
       "Q 76.703125 40.09375 76.703125 35.203125 \n",
       "L 76.703125 0 \n",
       "L 68.015625 0 \n",
       "L 68.015625 33.5 \n",
       "Q 68.015625 37.359375 67.53125 39.9375 \n",
       "Q 67.046875 42.53125 65.890625 44.109375 \n",
       "Q 64.75 45.703125 62.890625 46.359375 \n",
       "Q 61.03125 47.015625 58.34375 47.015625 \n",
       "Q 55.5625 47.015625 53.3125 45.96875 \n",
       "Q 51.078125 44.921875 49.484375 42.875 \n",
       "Q 47.90625 40.828125 47.046875 37.75 \n",
       "Q 46.1875 34.671875 46.1875 30.609375 \n",
       "L 46.1875 0 \n",
       "z\n",
       "\" id=\"LiberationSans-6d\"/>\n",
       "      <path d=\"M 51.421875 26.65625 \n",
       "Q 51.421875 -0.984375 31.984375 -0.984375 \n",
       "Q 25.984375 -0.984375 22 1.1875 \n",
       "Q 18.015625 3.375 15.53125 8.203125 \n",
       "L 15.4375 8.203125 \n",
       "Q 15.4375 6.9375 15.359375 5.5625 \n",
       "Q 15.28125 4.203125 15.203125 3.03125 \n",
       "Q 15.140625 1.859375 15.0625 1.03125 \n",
       "Q 14.984375 0.203125 14.9375 0 \n",
       "L 6.453125 0 \n",
       "Q 6.5 0.4375 6.546875 1.484375 \n",
       "Q 6.59375 2.546875 6.640625 4 \n",
       "Q 6.6875 5.46875 6.703125 7.21875 \n",
       "Q 6.734375 8.984375 6.734375 10.890625 \n",
       "L 6.734375 72.46875 \n",
       "L 15.53125 72.46875 \n",
       "L 15.53125 51.8125 \n",
       "Q 15.53125 50.34375 15.5 49 \n",
       "Q 15.484375 47.65625 15.4375 46.625 \n",
       "Q 15.375 45.40625 15.328125 44.34375 \n",
       "L 15.53125 44.34375 \n",
       "Q 17.96875 49.421875 22 51.609375 \n",
       "Q 26.03125 53.8125 31.984375 53.8125 \n",
       "Q 42 53.8125 46.703125 47.0625 \n",
       "Q 51.421875 40.328125 51.421875 26.65625 \n",
       "z\n",
       "M 42.1875 26.375 \n",
       "Q 42.1875 31.84375 41.5 35.75 \n",
       "Q 40.828125 39.65625 39.3125 42.15625 \n",
       "Q 37.796875 44.671875 35.453125 45.84375 \n",
       "Q 33.109375 47.015625 29.734375 47.015625 \n",
       "Q 26.265625 47.015625 23.609375 45.890625 \n",
       "Q 20.953125 44.78125 19.171875 42.28125 \n",
       "Q 17.390625 39.796875 16.453125 35.734375 \n",
       "Q 15.53125 31.6875 15.53125 25.828125 \n",
       "Q 15.53125 20.171875 16.453125 16.3125 \n",
       "Q 17.390625 12.453125 19.171875 10.03125 \n",
       "Q 20.953125 7.625 23.578125 6.5625 \n",
       "Q 26.21875 5.515625 29.640625 5.515625 \n",
       "Q 32.859375 5.515625 35.203125 6.640625 \n",
       "Q 37.546875 7.765625 39.109375 10.25 \n",
       "Q 40.671875 12.75 41.421875 16.71875 \n",
       "Q 42.1875 20.703125 42.1875 26.375 \n",
       "z\n",
       "\" id=\"LiberationSans-62\"/>\n",
       "      <path d=\"M 6.9375 0 \n",
       "L 6.9375 40.53125 \n",
       "Q 6.9375 42.1875 6.90625 43.921875 \n",
       "Q 6.890625 45.65625 6.828125 47.265625 \n",
       "Q 6.78125 48.875 6.734375 50.28125 \n",
       "Q 6.6875 51.703125 6.640625 52.828125 \n",
       "L 14.9375 52.828125 \n",
       "Q 14.984375 51.703125 15.0625 50.265625 \n",
       "Q 15.140625 48.828125 15.203125 47.3125 \n",
       "Q 15.28125 45.796875 15.296875 44.40625 \n",
       "Q 15.328125 43.015625 15.328125 42.046875 \n",
       "L 15.53125 42.046875 \n",
       "Q 16.453125 45.0625 17.5 47.28125 \n",
       "Q 18.5625 49.515625 19.96875 50.953125 \n",
       "Q 21.390625 52.390625 23.34375 53.09375 \n",
       "Q 25.296875 53.8125 28.078125 53.8125 \n",
       "Q 29.15625 53.8125 30.125 53.640625 \n",
       "Q 31.109375 53.46875 31.640625 53.328125 \n",
       "L 31.640625 45.265625 \n",
       "Q 30.765625 45.515625 29.59375 45.625 \n",
       "Q 28.421875 45.75 26.953125 45.75 \n",
       "Q 23.921875 45.75 21.796875 44.375 \n",
       "Q 19.671875 43.015625 18.328125 40.59375 \n",
       "Q 17 38.1875 16.359375 34.84375 \n",
       "Q 15.71875 31.5 15.71875 27.546875 \n",
       "L 15.71875 0 \n",
       "z\n",
       "\" id=\"LiberationSans-72\"/>\n",
       "      <path d=\"M 17.625 46.4375 \n",
       "L 17.625 0 \n",
       "L 8.84375 0 \n",
       "L 8.84375 46.4375 \n",
       "L 1.421875 46.4375 \n",
       "L 1.421875 52.828125 \n",
       "L 8.84375 52.828125 \n",
       "L 8.84375 58.796875 \n",
       "Q 8.84375 61.671875 9.375 64.140625 \n",
       "Q 9.90625 66.609375 11.34375 68.4375 \n",
       "Q 12.796875 70.265625 15.28125 71.3125 \n",
       "Q 17.78125 72.359375 21.734375 72.359375 \n",
       "Q 23.296875 72.359375 24.96875 72.21875 \n",
       "Q 26.65625 72.078125 27.9375 71.78125 \n",
       "L 27.9375 65.09375 \n",
       "Q 27.09375 65.234375 26 65.359375 \n",
       "Q 24.90625 65.484375 24.03125 65.484375 \n",
       "Q 22.078125 65.484375 20.828125 64.9375 \n",
       "Q 19.578125 64.40625 18.875 63.40625 \n",
       "Q 18.171875 62.40625 17.890625 60.9375 \n",
       "Q 17.625 59.46875 17.625 57.5625 \n",
       "L 17.625 52.828125 \n",
       "L 27.9375 52.828125 \n",
       "L 27.9375 46.4375 \n",
       "z\n",
       "\" id=\"LiberationSans-66\"/>\n",
       "      <path d=\"M 6.6875 64.0625 \n",
       "L 6.6875 72.46875 \n",
       "L 15.484375 72.46875 \n",
       "L 15.484375 64.0625 \n",
       "z\n",
       "M 6.6875 0 \n",
       "L 6.6875 52.828125 \n",
       "L 15.484375 52.828125 \n",
       "L 15.484375 0 \n",
       "z\n",
       "\" id=\"LiberationSans-69\"/>\n",
       "      <path d=\"M 26.765625 -20.75 \n",
       "Q 22.21875 -20.75 18.703125 -19.8125 \n",
       "Q 15.1875 -18.890625 12.6875 -17.15625 \n",
       "Q 10.203125 -15.4375 8.640625 -13.03125 \n",
       "Q 7.078125 -10.640625 6.390625 -7.71875 \n",
       "L 15.234375 -6.453125 \n",
       "Q 16.109375 -10.109375 19.109375 -12.078125 \n",
       "Q 22.125 -14.0625 27 -14.0625 \n",
       "Q 29.984375 -14.0625 32.421875 -13.234375 \n",
       "Q 34.859375 -12.40625 36.5625 -10.5625 \n",
       "Q 38.28125 -8.734375 39.203125 -5.796875 \n",
       "Q 40.140625 -2.875 40.140625 1.3125 \n",
       "L 40.140625 9.8125 \n",
       "L 40.046875 9.8125 \n",
       "Q 39.0625 7.8125 37.625 5.984375 \n",
       "Q 36.1875 4.15625 34.109375 2.734375 \n",
       "Q 32.03125 1.3125 29.296875 0.453125 \n",
       "Q 26.5625 -0.390625 23.046875 -0.390625 \n",
       "Q 18.015625 -0.390625 14.421875 1.296875 \n",
       "Q 10.84375 2.984375 8.5625 6.34375 \n",
       "Q 6.296875 9.71875 5.25 14.71875 \n",
       "Q 4.203125 19.734375 4.203125 26.3125 \n",
       "Q 4.203125 32.671875 5.25 37.75 \n",
       "Q 6.296875 42.828125 8.65625 46.359375 \n",
       "Q 11.03125 49.90625 14.8125 51.78125 \n",
       "Q 18.609375 53.65625 24.03125 53.65625 \n",
       "Q 29.640625 53.65625 33.765625 51.09375 \n",
       "Q 37.890625 48.53125 40.140625 43.796875 \n",
       "L 40.234375 43.796875 \n",
       "Q 40.234375 45.015625 40.296875 46.53125 \n",
       "Q 40.375 48.046875 40.453125 49.390625 \n",
       "Q 40.53125 50.734375 40.625 51.703125 \n",
       "Q 40.71875 52.6875 40.828125 52.828125 \n",
       "L 49.171875 52.828125 \n",
       "Q 49.125 52.390625 49.078125 51.34375 \n",
       "Q 49.03125 50.296875 48.96875 48.828125 \n",
       "Q 48.921875 47.359375 48.890625 45.578125 \n",
       "Q 48.875 43.796875 48.875 41.890625 \n",
       "L 48.875 1.515625 \n",
       "Q 48.875 -9.578125 43.421875 -15.15625 \n",
       "Q 37.984375 -20.75 26.765625 -20.75 \n",
       "z\n",
       "M 40.140625 26.421875 \n",
       "Q 40.140625 31.9375 38.9375 35.859375 \n",
       "Q 37.75 39.796875 35.796875 42.28125 \n",
       "Q 33.84375 44.78125 31.328125 45.953125 \n",
       "Q 28.8125 47.125 26.171875 47.125 \n",
       "Q 22.796875 47.125 20.375 45.953125 \n",
       "Q 17.96875 44.78125 16.375 42.265625 \n",
       "Q 14.796875 39.75 14.03125 35.8125 \n",
       "Q 13.28125 31.890625 13.28125 26.421875 \n",
       "Q 13.28125 20.703125 14.03125 16.8125 \n",
       "Q 14.796875 12.9375 16.359375 10.546875 \n",
       "Q 17.921875 8.15625 20.3125 7.125 \n",
       "Q 22.703125 6.109375 26.03125 6.109375 \n",
       "Q 28.65625 6.109375 31.171875 7.21875 \n",
       "Q 33.6875 8.34375 35.6875 10.78125 \n",
       "Q 37.703125 13.234375 38.921875 17.09375 \n",
       "Q 40.140625 20.953125 40.140625 26.421875 \n",
       "z\n",
       "\" id=\"LiberationSans-67\"/>\n",
       "      <path d=\"M 20.21875 -0.984375 \n",
       "Q 12.25 -0.984375 8.25 3.21875 \n",
       "Q 4.25 7.421875 4.25 14.75 \n",
       "Q 4.25 19.96875 6.21875 23.3125 \n",
       "Q 8.203125 26.65625 11.390625 28.5625 \n",
       "Q 14.59375 30.46875 18.6875 31.203125 \n",
       "Q 22.796875 31.9375 27.046875 32.03125 \n",
       "L 38.921875 32.234375 \n",
       "L 38.921875 35.109375 \n",
       "Q 38.921875 38.375 38.234375 40.671875 \n",
       "Q 37.546875 42.96875 36.125 44.375 \n",
       "Q 34.71875 45.796875 32.59375 46.453125 \n",
       "Q 30.46875 47.125 27.59375 47.125 \n",
       "Q 25.046875 47.125 23 46.75 \n",
       "Q 20.953125 46.390625 19.4375 45.4375 \n",
       "Q 17.921875 44.484375 16.984375 42.84375 \n",
       "Q 16.0625 41.21875 15.765625 38.71875 \n",
       "L 6.59375 39.546875 \n",
       "Q 7.078125 42.671875 8.4375 45.28125 \n",
       "Q 9.8125 47.90625 12.328125 49.796875 \n",
       "Q 14.84375 51.703125 18.625 52.75 \n",
       "Q 22.40625 53.8125 27.78125 53.8125 \n",
       "Q 37.75 53.8125 42.765625 49.234375 \n",
       "Q 47.796875 44.671875 47.796875 36.03125 \n",
       "L 47.796875 13.28125 \n",
       "Q 47.796875 9.375 48.828125 7.390625 \n",
       "Q 49.859375 5.421875 52.734375 5.421875 \n",
       "Q 53.46875 5.421875 54.203125 5.515625 \n",
       "Q 54.9375 5.609375 55.609375 5.765625 \n",
       "L 55.609375 0.296875 \n",
       "Q 53.953125 -0.09375 52.3125 -0.28125 \n",
       "Q 50.6875 -0.484375 48.828125 -0.484375 \n",
       "Q 46.34375 -0.484375 44.5625 0.171875 \n",
       "Q 42.78125 0.828125 41.65625 2.171875 \n",
       "Q 40.53125 3.515625 39.9375 5.484375 \n",
       "Q 39.359375 7.46875 39.203125 10.109375 \n",
       "L 38.921875 10.109375 \n",
       "Q 37.5 7.5625 35.8125 5.515625 \n",
       "Q 34.125 3.46875 31.875 2.03125 \n",
       "Q 29.640625 0.59375 26.78125 -0.1875 \n",
       "Q 23.921875 -0.984375 20.21875 -0.984375 \n",
       "z\n",
       "M 22.21875 5.609375 \n",
       "Q 26.421875 5.609375 29.5625 7.140625 \n",
       "Q 32.71875 8.6875 34.78125 11.078125 \n",
       "Q 36.859375 13.484375 37.890625 16.3125 \n",
       "Q 38.921875 19.140625 38.921875 21.734375 \n",
       "L 38.921875 26.078125 \n",
       "L 29.296875 25.875 \n",
       "Q 26.078125 25.828125 23.171875 25.40625 \n",
       "Q 20.265625 25 18.0625 23.78125 \n",
       "Q 15.875 22.5625 14.578125 20.359375 \n",
       "Q 13.28125 18.171875 13.28125 14.59375 \n",
       "Q 13.28125 10.296875 15.59375 7.953125 \n",
       "Q 17.921875 5.609375 22.21875 5.609375 \n",
       "z\n",
       "\" id=\"LiberationSans-61\"/>\n",
       "      <path d=\"M 27.046875 0.390625 \n",
       "Q 25.046875 -0.140625 22.96875 -0.453125 \n",
       "Q 20.90625 -0.78125 18.171875 -0.78125 \n",
       "Q 7.625 -0.78125 7.625 11.1875 \n",
       "L 7.625 46.4375 \n",
       "L 1.515625 46.4375 \n",
       "L 1.515625 52.828125 \n",
       "L 7.953125 52.828125 \n",
       "L 10.546875 64.65625 \n",
       "L 16.40625 64.65625 \n",
       "L 16.40625 52.828125 \n",
       "L 26.171875 52.828125 \n",
       "L 26.171875 46.4375 \n",
       "L 16.40625 46.4375 \n",
       "L 16.40625 13.09375 \n",
       "Q 16.40625 9.28125 17.640625 7.734375 \n",
       "Q 18.890625 6.203125 21.96875 6.203125 \n",
       "Q 23.25 6.203125 24.4375 6.390625 \n",
       "Q 25.640625 6.59375 27.046875 6.890625 \n",
       "z\n",
       "\" id=\"LiberationSans-74\"/>\n",
       "      <path d=\"M 40.28125 0 \n",
       "L 40.28125 33.5 \n",
       "Q 40.28125 37.359375 39.71875 39.9375 \n",
       "Q 39.15625 42.53125 37.890625 44.109375 \n",
       "Q 36.625 45.703125 34.546875 46.359375 \n",
       "Q 32.46875 47.015625 29.390625 47.015625 \n",
       "Q 26.265625 47.015625 23.75 45.921875 \n",
       "Q 21.234375 44.828125 19.453125 42.75 \n",
       "Q 17.671875 40.671875 16.6875 37.625 \n",
       "Q 15.71875 34.578125 15.71875 30.609375 \n",
       "L 15.71875 0 \n",
       "L 6.9375 0 \n",
       "L 6.9375 41.546875 \n",
       "Q 6.9375 43.21875 6.90625 45.046875 \n",
       "Q 6.890625 46.875 6.828125 48.5 \n",
       "Q 6.78125 50.140625 6.734375 51.3125 \n",
       "Q 6.6875 52.484375 6.640625 52.828125 \n",
       "L 14.9375 52.828125 \n",
       "Q 14.984375 52.59375 15.03125 51.515625 \n",
       "Q 15.09375 50.4375 15.15625 49.046875 \n",
       "Q 15.234375 47.65625 15.28125 46.21875 \n",
       "Q 15.328125 44.78125 15.328125 43.796875 \n",
       "L 15.484375 43.796875 \n",
       "Q 16.75 46.09375 18.265625 47.953125 \n",
       "Q 19.78125 49.8125 21.78125 51.09375 \n",
       "Q 23.78125 52.390625 26.359375 53.09375 \n",
       "Q 28.953125 53.8125 32.375 53.8125 \n",
       "Q 36.765625 53.8125 39.9375 52.734375 \n",
       "Q 43.109375 51.65625 45.15625 49.40625 \n",
       "Q 47.21875 47.171875 48.171875 43.625 \n",
       "Q 49.125 40.09375 49.125 35.203125 \n",
       "L 49.125 0 \n",
       "z\n",
       "\" id=\"LiberationSans-6e\"/>\n",
       "      <path d=\"M 46.390625 14.59375 \n",
       "Q 46.390625 10.890625 44.9375 7.984375 \n",
       "Q 43.5 5.078125 40.765625 3.09375 \n",
       "Q 38.03125 1.125 34.046875 0.0625 \n",
       "Q 30.078125 -0.984375 24.953125 -0.984375 \n",
       "Q 20.359375 -0.984375 16.671875 -0.265625 \n",
       "Q 12.984375 0.4375 10.203125 2 \n",
       "Q 7.421875 3.5625 5.53125 6.125 \n",
       "Q 3.65625 8.6875 2.78125 12.40625 \n",
       "L 10.546875 13.921875 \n",
       "Q 11.671875 9.671875 15.1875 7.6875 \n",
       "Q 18.703125 5.71875 24.953125 5.71875 \n",
       "Q 27.78125 5.71875 30.140625 6.109375 \n",
       "Q 32.515625 6.5 34.21875 7.453125 \n",
       "Q 35.9375 8.40625 36.890625 9.984375 \n",
       "Q 37.84375 11.578125 37.84375 13.921875 \n",
       "Q 37.84375 16.3125 36.71875 17.84375 \n",
       "Q 35.59375 19.390625 33.59375 20.40625 \n",
       "Q 31.59375 21.4375 28.734375 22.1875 \n",
       "Q 25.875 22.953125 22.46875 23.875 \n",
       "Q 19.28125 24.703125 16.15625 25.734375 \n",
       "Q 13.03125 26.765625 10.515625 28.4375 \n",
       "Q 8.015625 30.125 6.453125 32.609375 \n",
       "Q 4.890625 35.109375 4.890625 38.875 \n",
       "Q 4.890625 46.09375 10.03125 49.875 \n",
       "Q 15.1875 53.65625 25.046875 53.65625 \n",
       "Q 33.796875 53.65625 38.9375 50.578125 \n",
       "Q 44.09375 47.515625 45.453125 40.71875 \n",
       "L 37.546875 39.75 \n",
       "Q 37.109375 41.796875 35.9375 43.1875 \n",
       "Q 34.765625 44.578125 33.109375 45.4375 \n",
       "Q 31.453125 46.296875 29.375 46.65625 \n",
       "Q 27.296875 47.015625 25.046875 47.015625 \n",
       "Q 19.09375 47.015625 16.25 45.203125 \n",
       "Q 13.421875 43.40625 13.421875 39.75 \n",
       "Q 13.421875 37.59375 14.46875 36.203125 \n",
       "Q 15.53125 34.8125 17.40625 33.859375 \n",
       "Q 19.28125 32.90625 21.921875 32.203125 \n",
       "Q 24.5625 31.5 27.734375 30.71875 \n",
       "Q 29.828125 30.171875 32.03125 29.5625 \n",
       "Q 34.234375 28.953125 36.296875 28.09375 \n",
       "Q 38.375 27.25 40.203125 26.09375 \n",
       "Q 42.046875 24.953125 43.40625 23.34375 \n",
       "Q 44.78125 21.734375 45.578125 19.578125 \n",
       "Q 46.390625 17.4375 46.390625 14.59375 \n",
       "z\n",
       "\" id=\"LiberationSans-73\"/>\n",
       "     </defs>\n",
       "     <g style=\"fill:#262626;\" transform=\"translate(15.171563 168.491406)rotate(-90)scale(0.11 -0.11)\">\n",
       "      <use xlink:href=\"#LiberationSans-4e\"/>\n",
       "      <use x=\"72.216797\" xlink:href=\"#LiberationSans-75\"/>\n",
       "      <use x=\"127.832031\" xlink:href=\"#LiberationSans-6d\"/>\n",
       "      <use x=\"211.132812\" xlink:href=\"#LiberationSans-62\"/>\n",
       "      <use x=\"266.748047\" xlink:href=\"#LiberationSans-65\"/>\n",
       "      <use x=\"322.363281\" xlink:href=\"#LiberationSans-72\"/>\n",
       "      <use x=\"355.664062\" xlink:href=\"#LiberationSans-20\"/>\n",
       "      <use x=\"383.447266\" xlink:href=\"#LiberationSans-6f\"/>\n",
       "      <use x=\"439.0625\" xlink:href=\"#LiberationSans-66\"/>\n",
       "      <use x=\"466.845703\" xlink:href=\"#LiberationSans-20\"/>\n",
       "      <use x=\"494.628906\" xlink:href=\"#LiberationSans-6d\"/>\n",
       "      <use x=\"577.929688\" xlink:href=\"#LiberationSans-69\"/>\n",
       "      <use x=\"600.146484\" xlink:href=\"#LiberationSans-67\"/>\n",
       "      <use x=\"655.761719\" xlink:href=\"#LiberationSans-72\"/>\n",
       "      <use x=\"689.0625\" xlink:href=\"#LiberationSans-61\"/>\n",
       "      <use x=\"744.677734\" xlink:href=\"#LiberationSans-74\"/>\n",
       "      <use x=\"772.460938\" xlink:href=\"#LiberationSans-69\"/>\n",
       "      <use x=\"794.677734\" xlink:href=\"#LiberationSans-6f\"/>\n",
       "      <use x=\"850.292969\" xlink:href=\"#LiberationSans-6e\"/>\n",
       "      <use x=\"905.908203\" xlink:href=\"#LiberationSans-73\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path clip-path=\"url(#pce3d112b9a)\" d=\"M 54.794119 224.64 \n",
       "L 82.463541 224.64 \n",
       "L 82.463541 217.737143 \n",
       "L 54.794119 217.737143 \n",
       "z\n",
       "\" style=\"fill:#4c72b0;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path clip-path=\"url(#pce3d112b9a)\" d=\"M 89.380896 224.64 \n",
       "L 117.050318 224.64 \n",
       "L 117.050318 176.32 \n",
       "L 89.380896 176.32 \n",
       "z\n",
       "\" style=\"fill:#4c72b0;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path clip-path=\"url(#pce3d112b9a)\" d=\"M 123.967673 224.64 \n",
       "L 151.637095 224.64 \n",
       "L 151.637095 17.554286 \n",
       "L 123.967673 17.554286 \n",
       "z\n",
       "\" style=\"fill:#4c72b0;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path clip-path=\"url(#pce3d112b9a)\" d=\"M 158.55445 224.64 \n",
       "L 186.223871 224.64 \n",
       "L 186.223871 183.222857 \n",
       "L 158.55445 183.222857 \n",
       "z\n",
       "\" style=\"fill:#4c72b0;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_7\">\n",
       "    <path clip-path=\"url(#pce3d112b9a)\" d=\"M 193.141227 224.64 \n",
       "L 220.810648 224.64 \n",
       "L 220.810648 224.64 \n",
       "L 193.141227 224.64 \n",
       "z\n",
       "\" style=\"fill:#4c72b0;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_8\">\n",
       "    <path clip-path=\"url(#pce3d112b9a)\" d=\"M 227.728004 224.64 \n",
       "L 255.397425 224.64 \n",
       "L 255.397425 210.834286 \n",
       "L 227.728004 210.834286 \n",
       "z\n",
       "\" style=\"fill:#4c72b0;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_9\">\n",
       "    <path clip-path=\"url(#pce3d112b9a)\" d=\"M 262.31478 224.64 \n",
       "L 289.984202 224.64 \n",
       "L 289.984202 217.737143 \n",
       "L 262.31478 217.737143 \n",
       "z\n",
       "\" style=\"fill:#4c72b0;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_10\">\n",
       "    <path clip-path=\"url(#pce3d112b9a)\" d=\"M 296.901557 224.64 \n",
       "L 324.570979 224.64 \n",
       "L 324.570979 134.902857 \n",
       "L 296.901557 134.902857 \n",
       "z\n",
       "\" style=\"fill:#4c72b0;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_11\">\n",
       "    <path clip-path=\"url(#pce3d112b9a)\" d=\"M 331.488334 224.64 \n",
       "L 359.157756 224.64 \n",
       "L 359.157756 224.64 \n",
       "L 331.488334 224.64 \n",
       "z\n",
       "\" style=\"fill:#4c72b0;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_12\">\n",
       "    <path d=\"M 39.575938 224.64 \n",
       "L 39.575938 7.2 \n",
       "\" style=\"fill:none;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_13\">\n",
       "    <path d=\"M 374.375938 224.64 \n",
       "L 374.375938 7.2 \n",
       "\" style=\"fill:none;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_14\">\n",
       "    <path d=\"M 39.575938 224.64 \n",
       "L 374.375938 224.64 \n",
       "\" style=\"fill:none;\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_15\">\n",
       "    <path d=\"M 39.575938 7.2 \n",
       "L 374.375938 7.2 \n",
       "\" style=\"fill:none;\"/>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"pce3d112b9a\">\n",
       "   <rect height=\"217.44\" width=\"334.8\" x=\"39.575938\" y=\"7.2\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7ff55d2e1f98>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "node_count = np.zeros(ts.num_nodes)\n",
    "for migration in ts.migrations():\n",
    "    node_count[migration.node] += 1\n",
    "plt.bar(np.arange(ts.num_nodes), node_count)\n",
    "plt.xlabel(\"Node ID\")\n",
    "plt.ylabel(\"Number of migrations\");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that node zero experienced very few migrations before it ended up in deme 2, where it coalesced with 4 (which never migrated). Node 2, on the other hand migrated 30 times before it finally coalesced with 7 in deme 0.\n",
    "\n",
    "Other forms of migration are also possible between specific demes at specific times. These are one of a number of different demographic event, which are dealt with in the next section."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Demographic events\n",
    "\n",
    "Demographic events allow us to model more complex histories involving changes to the population structure over time. These are specified using the ``demographic_events`` parameter to ``simulate``. Each demographic event occurs at a specific time, and the list of events must be supplied in the order they occur (backwards in time). There are a number of different types of demographic event, which we examine in turn."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Migration rate change\n",
    "\n",
    "Migration rate change events allow us to update the migration rate matrix at some point in time. We can either update a single cell in the matrix or all (non-diagonal) entries at the same time.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs/><g id=\"lines\" stroke=\"black\"><line x1=\"50.0\" x2=\"50.0\" y1=\"176.94468609898897\" y2=\"20.0\"/><line x1=\"50.0\" x2=\"83.33333333333334\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"33.333333333333336\" x2=\"33.333333333333336\" y1=\"180.0\" y2=\"176.94468609898897\"/><line x1=\"33.333333333333336\" x2=\"50.0\" y1=\"176.94468609898897\" y2=\"176.94468609898897\"/><line x1=\"66.66666666666667\" x2=\"66.66666666666667\" y1=\"180.0\" y2=\"176.94468609898897\"/><line x1=\"66.66666666666667\" x2=\"50.0\" y1=\"176.94468609898897\" y2=\"176.94468609898897\"/><line x1=\"116.66666666666667\" x2=\"116.66666666666667\" y1=\"165.03563215458772\" y2=\"20.0\"/><line x1=\"116.66666666666667\" x2=\"83.33333333333334\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"100.0\" x2=\"100.0\" y1=\"180.0\" y2=\"165.03563215458772\"/><line x1=\"100.0\" x2=\"116.66666666666667\" y1=\"165.03563215458772\" y2=\"165.03563215458772\"/><line x1=\"133.33333333333334\" x2=\"133.33333333333334\" y1=\"180.0\" y2=\"165.03563215458772\"/><line x1=\"133.33333333333334\" x2=\"116.66666666666667\" y1=\"165.03563215458772\" y2=\"165.03563215458772\"/></g><g font-size=\"14\" text-anchor=\"start\"><text x=\"121.66666666666667\" y=\"160.03563215458772\">5 (t=2.0)</text></g><g font-size=\"14\" text-anchor=\"end\"><text x=\"45.0\" y=\"171.94468609898897\">4 (t=0.4)</text></g><g font-size=\"14\" text-anchor=\"middle\"><text x=\"83.33333333333334\" y=\"15.0\">6 (t=21.4)</text><text x=\"33.333333333333336\" y=\"200.0\">2</text><text x=\"66.66666666666667\" y=\"200.0\">3</text><text x=\"100.0\" y=\"200.0\">0</text><text x=\"133.33333333333334\" y=\"200.0\">1</text></g><circle cx=\"83.33333333333334\" cy=\"20.0\" fill=\"blue\" r=\"3\"/><circle cx=\"50.0\" cy=\"176.94468609898897\" fill=\"blue\" r=\"3\"/><circle cx=\"33.333333333333336\" cy=\"180.0\" fill=\"blue\" r=\"3\"/><circle cx=\"66.66666666666667\" cy=\"180.0\" fill=\"blue\" r=\"3\"/><circle cx=\"116.66666666666667\" cy=\"165.03563215458772\" fill=\"red\" r=\"3\"/><circle cx=\"100.0\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"133.33333333333334\" cy=\"180.0\" fill=\"red\" r=\"3\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"start\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"end\"/></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts  = msprime.simulate(\n",
    "    population_configurations=[\n",
    "        msprime.PopulationConfiguration(sample_size=2),\n",
    "        msprime.PopulationConfiguration(sample_size=2)],\n",
    "    demographic_events=[\n",
    "        msprime.MigrationRateChange(20, rate=1.0, matrix_index=(0, 1))],\n",
    "    random_seed=2)\n",
    "tree = ts.first()\n",
    "colour_map = {0:\"red\", 1:\"blue\"}\n",
    "node_colours = {u: colour_map[tree.population(u)] for u in tree.nodes()}\n",
    "node_labels = {\n",
    "    u: (str(u) if tree.is_sample(u) else \"{} (t={:.1f})\".format(u, tree.time(u))) \n",
    "    for u in tree.nodes()}\n",
    "SVG(tree.draw(node_colours=node_colours, node_labels=node_labels))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As before, the sample 0 and 1, and 2 and 3 coalesce quickly within their own populations. However, because the migration rate between the populations is zero these lineages are isolated and would never coalesce without some change in demography. The migration rate change event happens at time 20, resulting in node 5 migrating to deme 1 soon afterwards. The lineages then coalesce at time 21.44."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mass migration\n",
    "This class of event allows us to move some proportion of the lineages in one deme to another at a particular time. This allows us to model population splits and admixture events. \n",
    "\n",
    "Population splits occur when some fraction of the lineages from a population migrate to another population."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs/><g id=\"lines\" stroke=\"black\"><line x1=\"43.75\" x2=\"43.75\" y1=\"154.87210235067323\" y2=\"20.0\"/><line x1=\"43.75\" x2=\"81.25\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"25.0\" x2=\"25.0\" y1=\"180.0\" y2=\"154.87210235067323\"/><line x1=\"25.0\" x2=\"43.75\" y1=\"154.87210235067323\" y2=\"154.87210235067323\"/><line x1=\"62.5\" x2=\"62.5\" y1=\"174.48146590091346\" y2=\"154.87210235067323\"/><line x1=\"62.5\" x2=\"43.75\" y1=\"154.87210235067323\" y2=\"154.87210235067323\"/><line x1=\"50.0\" x2=\"50.0\" y1=\"180.0\" y2=\"174.48146590091346\"/><line x1=\"50.0\" x2=\"62.5\" y1=\"174.48146590091346\" y2=\"174.48146590091346\"/><line x1=\"75.0\" x2=\"75.0\" y1=\"180.0\" y2=\"174.48146590091346\"/><line x1=\"75.0\" x2=\"62.5\" y1=\"174.48146590091346\" y2=\"174.48146590091346\"/><line x1=\"118.75\" x2=\"118.75\" y1=\"145.98986283039352\" y2=\"20.0\"/><line x1=\"118.75\" x2=\"81.25\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"100.0\" x2=\"100.0\" y1=\"180.0\" y2=\"145.98986283039352\"/><line x1=\"100.0\" x2=\"118.75\" y1=\"145.98986283039352\" y2=\"145.98986283039352\"/><line x1=\"137.5\" x2=\"137.5\" y1=\"155.5542494139535\" y2=\"145.98986283039352\"/><line x1=\"137.5\" x2=\"118.75\" y1=\"145.98986283039352\" y2=\"145.98986283039352\"/><line x1=\"125.0\" x2=\"125.0\" y1=\"180.0\" y2=\"155.5542494139535\"/><line x1=\"125.0\" x2=\"137.5\" y1=\"155.5542494139535\" y2=\"155.5542494139535\"/><line x1=\"150.0\" x2=\"150.0\" y1=\"180.0\" y2=\"155.5542494139535\"/><line x1=\"150.0\" x2=\"137.5\" y1=\"155.5542494139535\" y2=\"155.5542494139535\"/></g><g font-size=\"14\" text-anchor=\"start\"><text x=\"67.5\" y=\"169.48146590091346\">6</text><text x=\"123.75\" y=\"140.98986283039352\">9 (t=3.64)</text><text x=\"142.5\" y=\"150.5542494139535\">7</text></g><g font-size=\"14\" text-anchor=\"end\"><text x=\"38.75\" y=\"149.87210235067323\">8 (t=2.69)</text></g><g font-size=\"14\" text-anchor=\"middle\"><text x=\"81.25\" y=\"15.0\">10 (t=17.15)</text><text x=\"25.0\" y=\"200.0\">2</text><text x=\"50.0\" y=\"200.0\">0</text><text x=\"75.0\" y=\"200.0\">1</text><text x=\"100.0\" y=\"200.0\">5</text><text x=\"125.0\" y=\"200.0\">3</text><text x=\"150.0\" y=\"200.0\">4</text></g><circle cx=\"81.25\" cy=\"20.0\" fill=\"red\" r=\"3\"/><circle cx=\"43.75\" cy=\"154.87210235067323\" fill=\"red\" r=\"3\"/><circle cx=\"25.0\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"62.5\" cy=\"174.48146590091346\" fill=\"red\" r=\"3\"/><circle cx=\"50.0\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"75.0\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"118.75\" cy=\"145.98986283039352\" fill=\"blue\" r=\"3\"/><circle cx=\"100.0\" cy=\"180.0\" fill=\"blue\" r=\"3\"/><circle cx=\"137.5\" cy=\"155.5542494139535\" fill=\"blue\" r=\"3\"/><circle cx=\"125.0\" cy=\"180.0\" fill=\"blue\" r=\"3\"/><circle cx=\"150.0\" cy=\"180.0\" fill=\"blue\" r=\"3\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"start\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"end\"/></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts  = msprime.simulate(\n",
    "    population_configurations=[\n",
    "        msprime.PopulationConfiguration(sample_size=3),\n",
    "        msprime.PopulationConfiguration(sample_size=3)],\n",
    "    demographic_events=[\n",
    "        msprime.MassMigration(15, source=1, dest=0, proportion=1)],\n",
    "    random_seed=20)\n",
    "tree = ts.first()\n",
    "colour_map = {0:\"red\", 1:\"blue\"}\n",
    "node_colours = {u: colour_map[tree.population(u)] for u in tree.nodes()}\n",
    "node_labels = {\n",
    "    u: (str(u) if u < 8 else \"{} (t={:.2f})\".format(u, tree.time(u))) \n",
    "    for u in tree.nodes()}\n",
    "SVG(tree.draw(node_colours=node_colours, node_labels=node_labels))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again, we have two isolated populations which coalesce down to a single lineage. The population split at time 15 (which produced all individuals in population 1), then results in this lineage migrating back to population 0, where it coalesces with the ancestor of the samples 0, 1 and 2.\n",
    "\n",
    "Admixture events (i.e., where some fraction of the lineages move to a different deme) are specified in the same way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs/><g id=\"lines\" stroke=\"black\"><line x1=\"25.0\" x2=\"25.0\" y1=\"180.0\" y2=\"20.0\"/><line x1=\"25.0\" x2=\"57.8125\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"90.625\" x2=\"90.625\" y1=\"76.76618649333746\" y2=\"20.0\"/><line x1=\"90.625\" x2=\"57.8125\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"62.5\" x2=\"62.5\" y1=\"177.25446495171246\" y2=\"76.76618649333746\"/><line x1=\"62.5\" x2=\"90.625\" y1=\"76.76618649333746\" y2=\"76.76618649333746\"/><line x1=\"50.0\" x2=\"50.0\" y1=\"180.0\" y2=\"177.25446495171246\"/><line x1=\"50.0\" x2=\"62.5\" y1=\"177.25446495171246\" y2=\"177.25446495171246\"/><line x1=\"75.0\" x2=\"75.0\" y1=\"180.0\" y2=\"177.25446495171246\"/><line x1=\"75.0\" x2=\"62.5\" y1=\"177.25446495171246\" y2=\"177.25446495171246\"/><line x1=\"118.75\" x2=\"118.75\" y1=\"164.5168942763958\" y2=\"76.76618649333746\"/><line x1=\"118.75\" x2=\"90.625\" y1=\"76.76618649333746\" y2=\"76.76618649333746\"/><line x1=\"100.0\" x2=\"100.0\" y1=\"180.0\" y2=\"164.5168942763958\"/><line x1=\"100.0\" x2=\"118.75\" y1=\"164.5168942763958\" y2=\"164.5168942763958\"/><line x1=\"137.5\" x2=\"137.5\" y1=\"178.8673780040899\" y2=\"164.5168942763958\"/><line x1=\"137.5\" x2=\"118.75\" y1=\"164.5168942763958\" y2=\"164.5168942763958\"/><line x1=\"125.0\" x2=\"125.0\" y1=\"180.0\" y2=\"178.8673780040899\"/><line x1=\"125.0\" x2=\"137.5\" y1=\"178.8673780040899\" y2=\"178.8673780040899\"/><line x1=\"150.0\" x2=\"150.0\" y1=\"180.0\" y2=\"178.8673780040899\"/><line x1=\"150.0\" x2=\"137.5\" y1=\"178.8673780040899\" y2=\"178.8673780040899\"/></g><g font-size=\"14\" text-anchor=\"start\"><text x=\"95.625\" y=\"71.76618649333746\">9 (t=4.5)</text><text x=\"123.75\" y=\"159.5168942763958\">8 (t=0.7)</text><text x=\"142.5\" y=\"173.8673780040899\">6</text></g><g font-size=\"14\" text-anchor=\"end\"><text x=\"57.5\" y=\"172.25446495171246\">7</text></g><g font-size=\"14\" text-anchor=\"middle\"><text x=\"57.8125\" y=\"15.0\">10 (t=6.9)</text><text x=\"25.0\" y=\"200.0\">5</text><text x=\"50.0\" y=\"200.0\">2</text><text x=\"75.0\" y=\"200.0\">4</text><text x=\"100.0\" y=\"200.0\">0</text><text x=\"125.0\" y=\"200.0\">1</text><text x=\"150.0\" y=\"200.0\">3</text></g><circle cx=\"57.8125\" cy=\"20.0\" fill=\"blue\" r=\"3\"/><circle cx=\"25.0\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"90.625\" cy=\"76.76618649333746\" fill=\"blue\" r=\"3\"/><circle cx=\"62.5\" cy=\"177.25446495171246\" fill=\"red\" r=\"3\"/><circle cx=\"50.0\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"75.0\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"118.75\" cy=\"164.5168942763958\" fill=\"blue\" r=\"3\"/><circle cx=\"100.0\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"137.5\" cy=\"178.8673780040899\" fill=\"red\" r=\"3\"/><circle cx=\"125.0\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"150.0\" cy=\"180.0\" fill=\"red\" r=\"3\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"start\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"end\"/></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts  = msprime.simulate(\n",
    "    population_configurations=[\n",
    "        msprime.PopulationConfiguration(sample_size=6),\n",
    "        msprime.PopulationConfiguration(sample_size=0)],\n",
    "    demographic_events=[\n",
    "        msprime.MassMigration(0.5, source=0, dest=1, proportion=0.5),\n",
    "        msprime.MigrationRateChange(1.1, rate=0.1),\n",
    "    ],\n",
    "    random_seed=26)\n",
    "tree = ts.first()\n",
    "colour_map = {0:\"red\", 1:\"blue\"}\n",
    "node_colours = {u: colour_map[tree.population(u)] for u in tree.nodes()}\n",
    "node_labels = {\n",
    "    u: (str(u) if u < 8 else \"{} (t={:.1f})\".format(u, tree.time(u))) \n",
    "    for u in tree.nodes()}\n",
    "SVG(tree.draw(node_colours=node_colours, node_labels=node_labels))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We begin in this example with 6 lineages sampled in population 0, zero samples in population 1, and with no migration between these populations. Then, at time 0.5, we have an admixture event where each of the 4 extant lineages (5, 7, 0 and 6) has a probability of 1/2 of moving to deme 1. Linages 0 and 6 migrate, and subsequently coalesce into node 8. Further back in time, at $t=1.1$, another demographic event occurs, changing the migration rate between the demes to 0.1, thereby allowing lineages to move between them. Eventually, all lineages end up in deme 1, where they coalesce into the MRCA at time $t=6.9$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population parameter change\n",
    "\n",
    "This class of event represents a simple change in the growth rate or size of a particular population. Since each population has it's own individual size and growth rates, we can change these arbitrarily as we go backwards in time. Keeping track of the actual sizes of different populations can be a little challenging, and for this reason msprime provides a ``DemographyDebugger`` class.\n",
    "\n",
    "To illustrate this we consider a very simple example in which we have a single population experiencing an phase of exponential growth from 750 to 100 generations ago. The size of the population 750 generations ago was 2000, and it grew to 20,000 over the next 650 generations. The size of the population has been stable at this value for the past 100 generations. We can encode this model as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=============================\n",
      "Epoch: 0 -- 100.0 generations\n",
      "=============================\n",
      "     start     end      growth_rate |     0    \n",
      "   -------- --------       -------- | -------- \n",
      "0 |  2e+04    2e+04               0 |     0    \n",
      "\n",
      "Events @ generation 100.0\n",
      "   - Population parameter change for -1: growth_rate -> 0.0035424386046062238 \n",
      "\n",
      "=================================\n",
      "Epoch: 100.0 -- 750.0 generations\n",
      "=================================\n",
      "     start     end      growth_rate |     0    \n",
      "   -------- --------       -------- | -------- \n",
      "0 |  2e+04    2e+03         0.00354 |     0    \n",
      "\n",
      "Events @ generation 750.0\n",
      "   - Population parameter change for -1: growth_rate -> 0 \n",
      "\n",
      "===============================\n",
      "Epoch: 750.0 -- inf generations\n",
      "===============================\n",
      "     start     end      growth_rate |     0    \n",
      "   -------- --------       -------- | -------- \n",
      "0 |  2e+03    2e+03               0 |     0    \n",
      "\n"
     ]
    }
   ],
   "source": [
    "N1 = 20000  # Population size at present\n",
    "N2 = 2000   # Population size at start (forwards in time) of exponential growth.\n",
    "T1 = 100    # End of exponential growth period (forwards in time)\n",
    "T2 = 750    # Start of exponential growth period (forwards in time) \n",
    "# Calculate growth rate; solve N2 = N1 * exp(-alpha * (T2 - T1))\n",
    "growth_rate = -np.log(N2 / N1) / (T2 - T1)\n",
    "population_configurations = [\n",
    "    msprime.PopulationConfiguration(sample_size=2, initial_size=N1)\n",
    "]\n",
    "demographic_events = [\n",
    "    msprime.PopulationParametersChange(time=T1, growth_rate=growth_rate),\n",
    "    msprime.PopulationParametersChange(time=T2, growth_rate=0),\n",
    "]\n",
    "dp = msprime.DemographyDebugger(           \n",
    "    population_configurations=population_configurations,\n",
    "    demographic_events=demographic_events)\n",
    "dp.print_history()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After we set up our model, we use the ``DemographyDebugger`` to check our calculations. We see that time has been split into three \"epochs\". From the present until 100 generations ago, we see that the population size is constant at 20,000. Then, we have a demographic event that changes the growth rate to 0.0035, which applies over the next epoch (from 100 to 750 generations ago). Over this time, the population grows from 2000 to 20,000, as we specified (note that the \"start\" and \"end\" of each epoch is looking *backwards* in time, as we consider epochs starting from the present and moving backwards). At generation 750 another event occurs, setting the growth rate for the population to 0. Then, the population size is constant at 20,000 from generation 750 until the indefinite past.\n",
    "\n",
    "A more complex example involving the three population out-of-Africa model is available in the online documentation.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ancient samples\n",
    "\n",
    "Up to this point we have assumed that all samples are taken at the present time. However, msprime allows us to specify abitrary sampling times and locations, allowing us to simulate (for example) ancient samples.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg baseProfile=\"full\" height=\"200\" version=\"1.1\" width=\"200\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:ev=\"http://www.w3.org/2001/xml-events\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"><defs/><g id=\"lines\" stroke=\"black\"><line x1=\"33.333333333333336\" x2=\"33.333333333333336\" y1=\"130.4988578854326\" y2=\"20.0\"/><line x1=\"33.333333333333336\" x2=\"62.5\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"91.66666666666667\" x2=\"91.66666666666667\" y1=\"91.4215836042686\" y2=\"20.0\"/><line x1=\"91.66666666666667\" x2=\"62.5\" y1=\"20.0\" y2=\"20.0\"/><line x1=\"66.66666666666667\" x2=\"66.66666666666667\" y1=\"180.0\" y2=\"91.4215836042686\"/><line x1=\"66.66666666666667\" x2=\"91.66666666666667\" y1=\"91.4215836042686\" y2=\"91.4215836042686\"/><line x1=\"116.66666666666667\" x2=\"116.66666666666667\" y1=\"169.71363880110022\" y2=\"91.4215836042686\"/><line x1=\"116.66666666666667\" x2=\"91.66666666666667\" y1=\"91.4215836042686\" y2=\"91.4215836042686\"/><line x1=\"100.0\" x2=\"100.0\" y1=\"180.0\" y2=\"169.71363880110022\"/><line x1=\"100.0\" x2=\"116.66666666666667\" y1=\"169.71363880110022\" y2=\"169.71363880110022\"/><line x1=\"133.33333333333334\" x2=\"133.33333333333334\" y1=\"180.0\" y2=\"169.71363880110022\"/><line x1=\"133.33333333333334\" x2=\"116.66666666666667\" y1=\"169.71363880110022\" y2=\"169.71363880110022\"/></g><g font-size=\"14\" text-anchor=\"start\"><text x=\"96.66666666666667\" y=\"86.4215836042686\">5</text><text x=\"121.66666666666667\" y=\"164.71363880110022\">4</text></g><g font-size=\"14\" text-anchor=\"end\"/><g font-size=\"14\" text-anchor=\"middle\"><text x=\"62.5\" y=\"15.0\">6</text><text x=\"33.333333333333336\" y=\"150.4988578854326\">3 (t=0.75)</text><text x=\"66.66666666666667\" y=\"200.0\">2</text><text x=\"100.0\" y=\"200.0\">0</text><text x=\"133.33333333333334\" y=\"200.0\">1</text></g><circle cx=\"62.5\" cy=\"20.0\" fill=\"blue\" r=\"3\"/><circle cx=\"33.333333333333336\" cy=\"130.4988578854326\" fill=\"blue\" r=\"3\"/><circle cx=\"91.66666666666667\" cy=\"91.4215836042686\" fill=\"blue\" r=\"3\"/><circle cx=\"66.66666666666667\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"116.66666666666667\" cy=\"169.71363880110022\" fill=\"red\" r=\"3\"/><circle cx=\"100.0\" cy=\"180.0\" fill=\"red\" r=\"3\"/><circle cx=\"133.33333333333334\" cy=\"180.0\" fill=\"red\" r=\"3\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"start\"/><g alignment-baseline=\"middle\" font-size=\"14\" font-style=\"italic\" text-anchor=\"end\"/></svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts = msprime.simulate(\n",
    "    samples=[\n",
    "        msprime.Sample(0, 0), msprime.Sample(0, 0), msprime.Sample(0, 0),\n",
    "        msprime.Sample(1, 0.75), # Ancient sample in deme 1\n",
    "    ],\n",
    "    population_configurations=[\n",
    "        msprime.PopulationConfiguration(),\n",
    "        msprime.PopulationConfiguration()],\n",
    "    migration_matrix=[\n",
    "        [0, 1],\n",
    "        [1, 0]],\n",
    "    random_seed=22)\n",
    "tree = ts.first()\n",
    "colour_map = {0:\"red\", 1:\"blue\"}\n",
    "node_colours = {u: colour_map[tree.population(u)] for u in tree.nodes()}\n",
    "node_labels = {\n",
    "    u: (str(u) if u != 3 else \"{} (t={:.2f})\".format(u, tree.time(u))) \n",
    "    for u in tree.nodes()}\n",
    "SVG(tree.draw(node_colours=node_colours, node_labels=node_labels))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**TODO** description of the tree."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Recombination\n",
    "\n",
    "How to specify recombination, and how to deal with the fact that we have a sequence of trees. Recombination maps. Also explain how we can get models with a fixed number of loci."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
